{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ictv-mmseqs2-protein-database\n",
    "\n",
    "This repository contains instructions to generate a MMSeqs2 protein database with ICTV taxonomy. This database was not benchmarked. For taxonomic assignment of viral genomes you can try [geNomad](https://github.com/apcamargo/genomad).\n",
    "\n",
    "### Dependencies:\n",
    "\n",
    "- [`aria2`](https://github.com/aria2/aria2)\n",
    "- [`ripgrep`](https://github.com/BurntSushi/ripgrep)\n",
    "- [`csvtk`](https://github.com/shenwei356/csvtk)\n",
    "- [`seqkit`](https://github.com/shenwei356/seqkit)\n",
    "- [`taxonkit` (version 0.11.1)](https://github.com/shenwei356/taxonkit/releases/tag/v0.11.1)\n",
    "- [`taxopy`](https://github.com/apcamargo/taxopy)\n",
    "- [`mmseqs2`](https://github.com/soedinglab/MMseqs2)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Instructions\n",
    "First we create an move to a comfortable working directory\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "os.makedirs(f\"/clusterfs/jgi/scratch/science/metagen/neri/projects/ictv_antonio\",exist_ok=True)\n",
    "os.chdir(f\"/clusterfs/jgi/scratch/science/metagen/neri/projects/ictv_antonio\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " Then, we create a conda enviroment for reproducibilty etc, and to install the dependencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "# # # Create and activate a new conda environment\n",
    "# mamba create -n ictv_mmseqs2 python=3.9 -y\n",
    "# mamba activate ictv_mmseqs2\n",
    "\n",
    "# # # Install conda packages\n",
    "# mamba install -c bioconda -c conda-forge \\\n",
    "#     aria2 \\\n",
    "#     ripgrep \\\n",
    "#     csvtk \\\n",
    "#     seqkit \\\n",
    "#     taxonkit=0.11.1 \\\n",
    "#     mmseqs2 \\\n",
    "#     polars\n",
    "#     -y\n",
    "# # ### taxonkit has to be 0.11.1 !!\n",
    "# # # Install taxopy using pip\n",
    "# pip install taxopy ipython jupyter\n",
    "# # # Note- now change your i/pythonkernel to the one you just created"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Verify installations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking installed tools:\n",
      "aria2 version 1.37.0\n",
      "ripgrep 14.1.0 (rev 9477456963)\n",
      "csvtk v0.30.0\n",
      "seqkit v2.8.2\n",
      "taxonkit v0.11.1\n",
      "taxopy 0.13.0\n",
      "15.6f452\n",
      "\n",
      "Setup complete. You can now run the rest of the thing.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "echo \"Checking installed tools:\"\n",
    "aria2c --version | head -n 1\n",
    "rg --version | head -n 1\n",
    "csvtk version | head -n 1\n",
    "seqkit version | head -n 1\n",
    "taxonkit version | head -n 1\n",
    "python -c \"import taxopy; print(f'taxopy {taxopy.__version__}')\"\n",
    "mmseqs version | head -n 1\n",
    "\n",
    "echo -e \"\\nSetup complete. You can now run the rest of the thing.\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, download the latest VMR release from ICTV, and refseq & genbank assembly status files from NCBI's FTP.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "aria2c -x 4 -o ictv.xlsx \"https://ictv.global/vmr/current?fid=15873\" --check-certificate=false\n",
    "aria2c -x 4 -o assembly_summary_genbank.tsv \"https://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_genbank.txt\" --check-certificate=false\n",
    "aria2c -x 4 -o assembly_summary_refseq.tsv \"https://ftp.ncbi.nlm.nih.gov/genomes/ASSEMBLY_REPORTS/assembly_summary_refseq.txt\" --check-certificate=false\n",
    "\n",
    "# wget https://ictv.global/vmr/current -O ictv.xlsx --no-check-certificate -o /dev/null\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### convert ICTV's xlsx to tsv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "csvtk xlsx2csv ictv.xlsx \\\n",
    "    | csvtk csv2tab \\\n",
    "    | sed 's/\\xc2\\xa0/ /g' \\\n",
    "    | csvtk replace -t -F -f \"*\" -p \"^\\s+|\\s+$\" \\\n",
    "    > ictv.tsv\n",
    "\n",
    "# choose columns, and remove duplicates\n",
    "csvtk cut -t -f \"Realm,Subrealm,Kingdom,Subkingdom,Phylum,Subphylum,Class,Subclass,Order,Suborder,Family,Subfamily,Genus,Subgenus,Species\" ictv.tsv \\\n",
    "    | csvtk uniq -t -f \"Realm,Subrealm,Kingdom,Subkingdom,Phylum,Subphylum,Class,Subclass,Order,Suborder,Family,Subfamily,Genus,Subgenus,Species\" \\\n",
    "    | csvtk del-header -t \\\n",
    "    > ictv.taxonomy.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a file that will store all the ICTV taxa names:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "csvtk cut -t -H -f 1,3,5,7,9,11,13,15 ictv.taxonomy.tsv \\\n",
    "    | sed 's/\\t/\\n/g' \\\n",
    "    | awk '!/^[[:blank:]]*$/' \\\n",
    "    | sort -u \\\n",
    "    > ictv.names.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's get NCBI taxdump "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "09/02 11:33:23 [\u001b[1;32mNOTICE\u001b[0m] Downloading 1 item(s)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[#82e959 0B/0B CN:1 DL:0B]\n",
      "\n",
      "09/02 11:33:25 [\u001b[1;32mNOTICE\u001b[0m] Allocating disk space. Use --file-allocation=none to disable it. See --file-allocation option in man page for more details.\n",
      "[#82e959 1.3MiB/63MiB(2%) CN:2 DL:3.6MiB ETA:17s]\n",
      "[#82e959 9.7MiB/63MiB(15%) CN:2 DL:7.1MiB ETA:7s]\n",
      "[#82e959 28MiB/63MiB(45%) CN:2 DL:12MiB ETA:2s]\n",
      "[#82e959 48MiB/63MiB(76%) CN:2 DL:14MiB ETA:1s]\n",
      "\n",
      "09/02 11:33:29 [\u001b[1;32mNOTICE\u001b[0m] Download complete: /clusterfs/jgi/scratch/science/metagen/neri/projects/ictv_antonio/taxdump.tar.gz\n",
      "\n",
      "Download Results:\n",
      "gid   |stat|avg speed  |path/URI\n",
      "======+====+===========+=======================================================\n",
      "82e959|OK  |    14MiB/s|/clusterfs/jgi/scratch/science/metagen/neri/projects/ictv_antonio/taxdump.tar.gz\n",
      "\n",
      "Status Legend:\n",
      "(OK):download completed.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "aria2c  -x 4 -o taxdump.tar.gz  ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz\n",
    "tar -xzf taxdump.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use `taxonkit create-taxdump` to create a custom taxdump for ICTV. Next, execute the `fix_taxdump.py` script, which will make the taxids sequential to make them compatible with MMSeqs2:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# mkdir ~/.taxonkit/\n",
    "# mv ./*dmp ~/.taxonkit/\n",
    "# mv gc.prt ~/.taxonkit/\n",
    "# mv readme* ~/.taxonkit/\n",
    "taxonkit create-taxdump -K 1 -P 3 -C 5 -O 7 -F 9 -G 11 -S 13 -T 15 \\\n",
    "    --rank-names \"realm\",\"kingdom\",\"phylum\",\"class\",\"order\",\"family\",\"genus\",\"species\" \\\n",
    "    ictv.taxonomy.tsv --out-dir ictv-taxdump --force\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This next block used to be called \"fix_taxdump.py\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "taxid_mapping_dict = {}\n",
    "count = 1\n",
    "with open(Path(\"ictv-taxdump\").joinpath(\"names.dmp\")) as fin:\n",
    "    for line in fin:\n",
    "        taxid = line.split(\"\\t\")[0]\n",
    "        taxid_mapping_dict[taxid] = str(count)\n",
    "        count += 1\n",
    "\n",
    "with open(Path(\"ictv-taxdump\").joinpath(\"names.dmp\")) as fin, open(\n",
    "    Path(\"ictv-taxdump\").joinpath(\"names.dmp.new\"), \"w\"\n",
    ") as fout:\n",
    "    for line in fin:\n",
    "        line = line.strip().split(\"\\t\")\n",
    "        line[0] = taxid_mapping_dict[line[0]]\n",
    "        line = \"\\t\".join(line)\n",
    "        fout.write(f\"{line}\\n\")\n",
    "\n",
    "with open(Path(\"ictv-taxdump\").joinpath(\"nodes.dmp\")) as fin, open(\n",
    "    Path(\"ictv-taxdump\").joinpath(\"nodes.dmp.new\"), \"w\"\n",
    ") as fout:\n",
    "    for line in fin:\n",
    "        line = line.strip().split(\"\\t\")\n",
    "        line[0] = taxid_mapping_dict[line[0]]\n",
    "        line[2] = taxid_mapping_dict[line[2]]\n",
    "        line = \"\\t\".join(line)\n",
    "        fout.write(f\"{line}\\n\")\n",
    "\n",
    "Path(\"ictv-taxdump\").joinpath(\"names.dmp\").unlink()\n",
    "Path(\"ictv-taxdump\").joinpath(\"nodes.dmp\").unlink()\n",
    "Path(\"ictv-taxdump\").joinpath(\"nodes.dmp.new\").rename(\"ictv-taxdump/nodes.dmp\")\n",
    "Path(\"ictv-taxdump\").joinpath(\"names.dmp.new\").rename(\"ictv-taxdump/names.dmp\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download the NCBI taxdump (again?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# Download the NCBI taxdump\n",
    "aria2c -x 4 \"ftp://ftp.ncbi.nih.gov/pub/taxonomy/taxdump.tar.gz\" --check-certificate=false #already there\n",
    "mkdir ncbi-taxdump\n",
    "tar zxfv taxdump.tar.gz -C ncbi-taxdump\n",
    "rm taxdump.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Download the protein → taxid association and filter for viruses\n",
    " via the `prot.accession2taxid` file filtered to keep only viral proteins:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "aria2c -x 4 \"https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/accession2taxid/prot.accession2taxid.FULL.gz\" --check-certificate=false"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl\n",
    "import requests\n",
    "from concurrent.futures import ThreadPoolExecutor\n",
    "import os\n",
    "\n",
    "# Function to download a specific file from the NCBI's assembly FTP directory\n",
    "def download_protein_file(row):\n",
    "    url = row['ftp_path']\n",
    "    taxid = str(row['taxid'])\n",
    "    file_name = os.path.basename(url.rstrip('/')) + '_protein.faa.gz'\n",
    "    download_url = os.path.join(url, file_name)\n",
    "\n",
    "    dest_folder = os.path.join('proteins', taxid)\n",
    "    local_filename = os.path.join(dest_folder, file_name)\n",
    "\n",
    "    os.makedirs(dest_folder, exist_ok=True)\n",
    "\n",
    "    try:\n",
    "        with requests.get(download_url, stream=True) as r:\n",
    "            r.raise_for_status()\n",
    "            with open(local_filename, 'wb') as f:\n",
    "                for chunk in r.iter_content(chunk_size=8192):\n",
    "                    f.write(chunk)\n",
    "        print(f\"Downloaded {download_url} to {local_filename}\")\n",
    "    except Exception as e:\n",
    "        print(f\"Failed to download {download_url}: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " Read the TSV files, filter to only viruses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (1, 2)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>group</th><th>count</th></tr><tr><td>str</td><td>u32</td></tr></thead><tbody><tr><td>&quot;viral&quot;</td><td>205904</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (1, 2)\n",
       "┌───────┬────────┐\n",
       "│ group ┆ count  │\n",
       "│ ---   ┆ ---    │\n",
       "│ str   ┆ u32    │\n",
       "╞═══════╪════════╡\n",
       "│ viral ┆ 205904 │\n",
       "└───────┴────────┘"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "refseq_df = pl.read_csv(\"assembly_summary_refseq.tsv\", separator='\\t',comment_prefix=\"##\",infer_schema_length=100000,has_header=True)\n",
    "genbank_df = pl.read_csv(\"assembly_summary_genbank.tsv\", separator='\\t',comment_prefix=\"##\",infer_schema_length=100000,has_header=True)\n",
    "\n",
    "# Concatenate the DataFrames\n",
    "df = pl.concat([refseq_df, genbank_df])\n",
    "\n",
    "# Filter for rows where the \"group\" column is \"viral\"\n",
    "df_viral = df.filter(pl.col('group') == 'viral')\n",
    "df_viral[\"group\"].value_counts()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use ThreadPoolExecutor to download with n threads at a time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_threads = 5  # Adjust as needed, more is gooder?\n",
    "with ThreadPoolExecutor(max_workers=n_threads) as executor:\n",
    "    executor.map(download_protein_file, df_viral.to_dicts())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Decompress and concatenate all proteins and create accession2taxid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import glob\n",
    "import gzip\n",
    "\n",
    "# Function to concatenate all protein FASTA files into a single file\n",
    "def concatenate_fasta_files(proteins_dir, output_fasta):\n",
    "    with open(output_fasta, 'w') as outfile:\n",
    "        for file_path in glob.glob(os.path.join(proteins_dir, '**/*.faa.gz'), recursive=True):\n",
    "            with gzip.open(file_path, 'rt') as infile:  # Decompress on the fly\n",
    "                outfile.write(infile.read())\n",
    "    print(f\"Concatenated all proteins into {output_fasta}\")\n",
    "\n",
    "# Function to create the virus.accession2taxid file - seperated from above as it might change\n",
    "def create_accession2taxid(proteins_dir, output_tsv):\n",
    "    with open(output_tsv, 'w') as outfile:\n",
    "        for file_path in glob.glob(os.path.join(proteins_dir, '**/*.faa.gz'), recursive=True):\n",
    "            taxid = os.path.basename(os.path.dirname(file_path))\n",
    "            with gzip.open(file_path, 'rt') as infile:  # Decompress on the fly\n",
    "                for line in infile:\n",
    "                    if line.startswith('>'):  # Find FASTA header lines\n",
    "                        accession = line.split()[0][1:]  # Extract accession (remove '>' and get until first space) - not sure if the defline/comment is to be expected too.\n",
    "                        outfile.write(f\"{accession}\\t{taxid}\\n\")\n",
    "    print(f\"Created virus.accession2taxid file at {output_tsv}\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " Define the directory containing the protein files and the output files\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Concatenated all proteins into concatenated_proteins.fasta\n",
      "Created virus.accession2taxid file at virus.accession2taxid.tsv\n"
     ]
    }
   ],
   "source": [
    "proteins_dir = './proteins'\n",
    "output_fasta = 'concatenated_proteins.fasta'\n",
    "output_tsv = 'virus.accession2taxid.tsv'\n",
    "\n",
    "# Concatenate all protein FASTA files into a single FASTA file\n",
    "concatenate_fasta_files(proteins_dir, output_fasta)\n",
    "\n",
    "# Create the virus.accession2taxid TSV file\n",
    "create_accession2taxid(proteins_dir, output_tsv) # from the NCBI data!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "18:22:48.717 \u001b[33m[WARN]\u001b[0m taxid 0 not found\n",
      "18:22:48.733 \u001b[33m[WARN]\u001b[0m taxid 1003800 was merged into 3053179\n",
      "18:22:48.754 \u001b[33m[WARN]\u001b[0m taxid 1035807 was merged into 2754702\n",
      "18:22:48.784 \u001b[33m[WARN]\u001b[0m taxid 1051812 was merged into 2493788\n",
      "18:22:48.788 \u001b[33m[WARN]\u001b[0m taxid 1053761 was merged into 2749666\n",
      "18:22:48.807 \u001b[33m[WARN]\u001b[0m taxid 1071652 was merged into 3150596\n",
      "18:22:48.840 \u001b[33m[WARN]\u001b[0m taxid 1097393 was merged into 3150328\n",
      "18:22:48.849 \u001b[33m[WARN]\u001b[0m taxid 1105366 was merged into 3149532\n",
      "18:22:48.854 \u001b[33m[WARN]\u001b[0m taxid 11103 was merged into 3052230\n",
      "18:22:48.863 \u001b[33m[WARN]\u001b[0m taxid 1118840 was merged into 3150192\n",
      "18:22:48.866 \u001b[33m[WARN]\u001b[0m taxid 11232 was merged into 3052342\n",
      "18:22:48.895 \u001b[33m[WARN]\u001b[0m taxid 1149653 was merged into 3150600\n",
      "18:22:48.898 \u001b[33m[WARN]\u001b[0m taxid 1154709 was merged into 565572\n",
      "18:22:48.900 \u001b[33m[WARN]\u001b[0m taxid 1156350 was merged into 3151018\n",
      "18:22:48.902 \u001b[33m[WARN]\u001b[0m taxid 1158301 was merged into 1653868\n",
      "18:22:48.965 \u001b[33m[WARN]\u001b[0m taxid 1237105 was merged into 702568\n",
      "18:22:48.972 \u001b[33m[WARN]\u001b[0m taxid 1243022 was merged into 1381081\n",
      "18:22:49.005 \u001b[33m[WARN]\u001b[0m taxid 1262729 was merged into 2981915\n",
      "18:22:49.015 \u001b[33m[WARN]\u001b[0m taxid 126877 was merged into 3150913\n",
      "18:22:49.015 \u001b[33m[WARN]\u001b[0m taxid 126912 was merged into 1689810\n",
      "18:22:49.031 \u001b[33m[WARN]\u001b[0m taxid 128078 was merged into 3315561\n",
      "18:22:49.039 \u001b[33m[WARN]\u001b[0m taxid 1285098 was merged into 3122678\n",
      "18:22:49.057 \u001b[33m[WARN]\u001b[0m taxid 1297391 was merged into 499957\n",
      "18:22:49.101 \u001b[33m[WARN]\u001b[0m taxid 1316683 was merged into 1316417\n",
      "18:22:49.144 \u001b[33m[WARN]\u001b[0m taxid 1345260 was merged into 3076337\n",
      "18:22:49.153 \u001b[33m[WARN]\u001b[0m taxid 135888 was merged into 1553355\n",
      "18:22:49.183 \u001b[33m[WARN]\u001b[0m taxid 137779 was merged into 184924\n",
      "18:22:49.247 \u001b[33m[WARN]\u001b[0m taxid 140663 was merged into 2854737\n",
      "18:22:49.247 \u001b[33m[WARN]\u001b[0m taxid 1407019 was merged into 555512\n",
      "18:22:49.354 \u001b[33m[WARN]\u001b[0m taxid 1451140 was merged into 3238682\n",
      "18:22:49.362 \u001b[33m[WARN]\u001b[0m taxid 1454568 was merged into 393709\n",
      "18:22:49.402 \u001b[33m[WARN]\u001b[0m taxid 1463951 was merged into 3151024\n",
      "18:22:49.403 \u001b[33m[WARN]\u001b[0m taxid 1463952 was merged into 3151033\n",
      "18:22:49.403 \u001b[33m[WARN]\u001b[0m taxid 1464012 was merged into 1220172\n",
      "18:22:49.403 \u001b[33m[WARN]\u001b[0m taxid 1464015 was merged into 3150401\n",
      "18:22:49.403 \u001b[33m[WARN]\u001b[0m taxid 1464026 was merged into 3149876\n",
      "18:22:49.403 \u001b[33m[WARN]\u001b[0m taxid 1464067 was merged into 3149433\n",
      "18:22:49.489 \u001b[33m[WARN]\u001b[0m taxid 1480111 was merged into 3248526\n",
      "18:22:49.723 \u001b[33m[WARN]\u001b[0m taxid 1500675 was merged into 2691557\n",
      "18:22:49.723 \u001b[33m[WARN]\u001b[0m taxid 1500676 was merged into 2691560\n",
      "18:22:49.728 \u001b[33m[WARN]\u001b[0m taxid 1503070 was merged into 3252912\n",
      "18:22:49.752 \u001b[33m[WARN]\u001b[0m taxid 1517773 was merged into 3150712\n",
      "18:22:49.752 \u001b[33m[WARN]\u001b[0m taxid 1517783 was merged into 3149765\n",
      "18:22:49.753 \u001b[33m[WARN]\u001b[0m taxid 1517779 was merged into 1517808\n",
      "18:22:49.753 \u001b[33m[WARN]\u001b[0m taxid 1517788 was merged into 1517809\n",
      "18:22:49.753 \u001b[33m[WARN]\u001b[0m taxid 1517803 was merged into 3150121\n",
      "18:22:49.753 \u001b[33m[WARN]\u001b[0m taxid 1517829 was merged into 3149766\n",
      "18:22:49.756 \u001b[33m[WARN]\u001b[0m taxid 1520093 was merged into 2547890\n",
      "18:22:49.823 \u001b[33m[WARN]\u001b[0m taxid 1550026 was merged into 1123738\n",
      "18:22:49.886 \u001b[33m[WARN]\u001b[0m taxid 158080 was merged into 141390\n",
      "18:22:49.915 \u001b[33m[WARN]\u001b[0m taxid 1589741 was merged into 171406\n",
      "18:22:49.965 \u001b[33m[WARN]\u001b[0m taxid 1608427 was merged into 3149841\n",
      "18:22:50.028 \u001b[33m[WARN]\u001b[0m taxid 163842 was merged into 2483883\n",
      "18:22:50.240 \u001b[33m[WARN]\u001b[0m taxid 1730500 was merged into 2497753\n",
      "18:22:50.300 \u001b[33m[WARN]\u001b[0m taxid 176056 was merged into 3149821\n",
      "18:22:50.323 \u001b[33m[WARN]\u001b[0m taxid 1774956 was merged into 1703256\n",
      "18:22:50.349 \u001b[33m[WARN]\u001b[0m taxid 178697 was merged into 2780052\n",
      "18:22:50.349 \u001b[33m[WARN]\u001b[0m taxid 178700 was merged into 2806833\n",
      "18:22:50.352 \u001b[33m[WARN]\u001b[0m taxid 179107 was merged into 178379\n",
      "18:22:50.355 \u001b[33m[WARN]\u001b[0m taxid 179411 was merged into 2713617\n",
      "18:22:50.387 \u001b[33m[WARN]\u001b[0m taxid 181628 was merged into 3149731\n",
      "18:22:50.392 \u001b[33m[WARN]\u001b[0m taxid 181927 was merged into 2970140\n",
      "18:22:50.452 \u001b[33m[WARN]\u001b[0m taxid 186229 was merged into 3149491\n",
      "18:22:50.474 \u001b[33m[WARN]\u001b[0m taxid 187437 was merged into 3150915\n",
      "18:22:50.490 \u001b[33m[WARN]\u001b[0m taxid 187986 was merged into 1880901\n",
      "18:22:50.571 \u001b[33m[WARN]\u001b[0m taxid 1898638 was merged into 2807124\n",
      "18:22:50.640 \u001b[33m[WARN]\u001b[0m taxid 1906744 was merged into 1796644\n",
      "18:22:50.693 \u001b[33m[WARN]\u001b[0m taxid 1915243 was merged into 1830\n",
      "18:22:50.738 \u001b[33m[WARN]\u001b[0m taxid 1934148 was merged into 980145\n",
      "18:22:50.738 \u001b[33m[WARN]\u001b[0m taxid 1934149 was merged into 980146\n",
      "18:22:50.738 \u001b[33m[WARN]\u001b[0m taxid 1934150 was merged into 980147\n",
      "18:22:50.738 \u001b[33m[WARN]\u001b[0m taxid 1934181 was merged into 980166\n",
      "18:22:50.738 \u001b[33m[WARN]\u001b[0m taxid 1934122 was merged into 980130\n",
      "18:22:50.738 \u001b[33m[WARN]\u001b[0m taxid 1934123 was merged into 980131\n",
      "18:22:50.738 \u001b[33m[WARN]\u001b[0m taxid 1934124 was merged into 980132\n",
      "18:22:50.738 \u001b[33m[WARN]\u001b[0m taxid 1934125 was merged into 980133\n",
      "18:22:50.738 \u001b[33m[WARN]\u001b[0m taxid 1934126 was merged into 980134\n",
      "18:22:50.738 \u001b[33m[WARN]\u001b[0m taxid 1934128 was merged into 980135\n",
      "18:22:50.902 \u001b[33m[WARN]\u001b[0m taxid 201309 was merged into 3151016\n",
      "18:22:50.903 \u001b[33m[WARN]\u001b[0m taxid 201349 was merged into 2944899\n",
      "18:22:50.972 \u001b[33m[WARN]\u001b[0m taxid 2032624 was merged into 272774\n",
      "18:22:50.992 \u001b[33m[WARN]\u001b[0m taxid 2042705 was merged into 3273800\n",
      "18:22:51.000 \u001b[33m[WARN]\u001b[0m taxid 204789 was merged into 3150510\n",
      "18:22:51.000 \u001b[33m[WARN]\u001b[0m taxid 204790 was merged into 2896652\n",
      "18:22:51.023 \u001b[33m[WARN]\u001b[0m taxid 2054477 was merged into 669621\n",
      "18:22:51.061 \u001b[33m[WARN]\u001b[0m taxid 206428 was merged into 3336981\n",
      "18:22:51.101 \u001b[33m[WARN]\u001b[0m taxid 2078707 was merged into 3101731\n",
      "18:22:51.102 \u001b[33m[WARN]\u001b[0m taxid 2078709 was merged into 1079125\n",
      "18:22:51.142 \u001b[33m[WARN]\u001b[0m taxid 2093792 was merged into 3121606\n",
      "18:22:51.307 \u001b[33m[WARN]\u001b[0m taxid 216218 was merged into 3149720\n",
      "18:22:51.356 \u001b[33m[WARN]\u001b[0m taxid 219518 was merged into 1977147\n",
      "18:22:51.394 \u001b[33m[WARN]\u001b[0m taxid 2202229 was merged into 3297708\n",
      "18:22:51.693 \u001b[33m[WARN]\u001b[0m taxid 225471 was merged into 1843498\n",
      "18:22:51.703 \u001b[33m[WARN]\u001b[0m taxid 2259161 was merged into 268010\n",
      "18:22:51.743 \u001b[33m[WARN]\u001b[0m taxid 2282268 was merged into 1913622\n",
      "18:22:51.743 \u001b[33m[WARN]\u001b[0m taxid 2282281 was merged into 2912030\n",
      "18:22:51.744 \u001b[33m[WARN]\u001b[0m taxid 228257 was merged into 3082113\n",
      "18:22:51.807 \u001b[33m[WARN]\u001b[0m taxid 23094 was merged into 2807387\n",
      "18:22:51.941 \u001b[33m[WARN]\u001b[0m taxid 2358191 was merged into 3150982\n",
      "18:22:52.143 \u001b[33m[WARN]\u001b[0m taxid 240713 was merged into 3150392\n",
      "18:22:52.143 \u001b[33m[WARN]\u001b[0m taxid 240716 was merged into 2585815\n",
      "18:22:52.143 \u001b[33m[WARN]\u001b[0m taxid 240715 was merged into 3150396\n",
      "18:22:52.642 \u001b[33m[WARN]\u001b[0m taxid 2482795 was merged into 2852101\n",
      "18:22:52.642 \u001b[33m[WARN]\u001b[0m taxid 2482801 was merged into 2818158\n",
      "18:22:52.656 \u001b[33m[WARN]\u001b[0m taxid 2487701 was merged into 2742613\n",
      "18:22:52.656 \u001b[33m[WARN]\u001b[0m taxid 2487702 was merged into 2742612\n",
      "18:22:52.656 \u001b[33m[WARN]\u001b[0m taxid 2487703 was merged into 2742615\n",
      "18:22:52.693 \u001b[33m[WARN]\u001b[0m taxid 2498560 was merged into 3243661\n",
      "18:22:52.859 \u001b[33m[WARN]\u001b[0m taxid 2582834 was merged into 2582835\n",
      "18:22:52.885 \u001b[33m[WARN]\u001b[0m taxid 2597149 was merged into 3273799\n",
      "18:22:52.888 \u001b[33m[WARN]\u001b[0m taxid 259884 was merged into 528016\n",
      "18:22:52.890 \u001b[33m[WARN]\u001b[0m taxid 260102 was merged into 3150080\n",
      "18:22:52.912 \u001b[33m[WARN]\u001b[0m taxid 262131 was merged into 2953425\n",
      "18:22:52.914 \u001b[33m[WARN]\u001b[0m taxid 263867 was merged into 3273798\n",
      "18:22:52.914 \u001b[33m[WARN]\u001b[0m taxid 263868 was merged into 3273801\n",
      "18:22:52.915 \u001b[33m[WARN]\u001b[0m taxid 264447 was merged into 3274530\n",
      "18:22:52.928 \u001b[33m[WARN]\u001b[0m taxid 265977 was merged into 326384\n",
      "18:22:52.928 \u001b[33m[WARN]\u001b[0m taxid 265982 was merged into 225452\n",
      "18:22:52.929 \u001b[33m[WARN]\u001b[0m taxid 266018 was merged into 2651253\n",
      "18:22:52.946 \u001b[33m[WARN]\u001b[0m taxid 2675855 was merged into 3038979\n",
      "18:22:52.947 \u001b[33m[WARN]\u001b[0m taxid 2675856 was merged into 3038979\n",
      "18:22:52.954 \u001b[33m[WARN]\u001b[0m taxid 2684394 was merged into 2662364\n",
      "18:22:52.975 \u001b[33m[WARN]\u001b[0m taxid 2705428 was merged into 2675321\n",
      "18:22:52.975 \u001b[33m[WARN]\u001b[0m taxid 2705429 was merged into 2675322\n",
      "18:22:52.975 \u001b[33m[WARN]\u001b[0m taxid 2705430 was merged into 2675323\n",
      "18:22:52.975 \u001b[33m[WARN]\u001b[0m taxid 2705431 was merged into 2675324\n",
      "18:22:52.975 \u001b[33m[WARN]\u001b[0m taxid 2705432 was merged into 2675325\n",
      "18:22:53.055 \u001b[33m[WARN]\u001b[0m taxid 274620 was merged into 3150328\n",
      "18:22:53.069 \u001b[33m[WARN]\u001b[0m taxid 2755037 was merged into 311194\n",
      "18:22:53.071 \u001b[33m[WARN]\u001b[0m taxid 2756263 was merged into 34022\n",
      "18:22:53.118 \u001b[33m[WARN]\u001b[0m taxid 2789583 was merged into 2841044\n",
      "18:22:53.121 \u001b[33m[WARN]\u001b[0m taxid 2791187 was deleted\n",
      "18:22:53.146 \u001b[33m[WARN]\u001b[0m taxid 2803558 was merged into 2871171\n",
      "18:22:53.217 \u001b[33m[WARN]\u001b[0m taxid 2820883 was merged into 2820060\n",
      "18:22:53.217 \u001b[33m[WARN]\u001b[0m taxid 2820879 was merged into 2820060\n",
      "18:22:53.218 \u001b[33m[WARN]\u001b[0m taxid 2820880 was merged into 2820060\n",
      "18:22:53.218 \u001b[33m[WARN]\u001b[0m taxid 2820881 was merged into 2820060\n",
      "18:22:53.218 \u001b[33m[WARN]\u001b[0m taxid 2820882 was merged into 2820060\n",
      "18:22:53.223 \u001b[33m[WARN]\u001b[0m taxid 2821276 was merged into 3003247\n",
      "18:22:53.231 \u001b[33m[WARN]\u001b[0m taxid 2821965 was deleted\n",
      "18:22:53.231 \u001b[33m[WARN]\u001b[0m taxid 2821966 was deleted\n",
      "18:22:53.231 \u001b[33m[WARN]\u001b[0m taxid 2821967 was deleted\n",
      "18:22:53.231 \u001b[33m[WARN]\u001b[0m taxid 2821968 was deleted\n",
      "18:22:53.231 \u001b[33m[WARN]\u001b[0m taxid 2821969 was deleted\n",
      "18:22:53.231 \u001b[33m[WARN]\u001b[0m taxid 2821970 was deleted\n",
      "18:22:53.232 \u001b[33m[WARN]\u001b[0m taxid 2821971 was deleted\n",
      "18:22:53.232 \u001b[33m[WARN]\u001b[0m taxid 2821972 was deleted\n",
      "18:22:53.232 \u001b[33m[WARN]\u001b[0m taxid 2821973 was deleted\n",
      "18:22:53.232 \u001b[33m[WARN]\u001b[0m taxid 2821976 was deleted\n",
      "18:22:53.232 \u001b[33m[WARN]\u001b[0m taxid 2821977 was deleted\n",
      "18:22:53.232 \u001b[33m[WARN]\u001b[0m taxid 2821978 was deleted\n",
      "18:22:53.232 \u001b[33m[WARN]\u001b[0m taxid 2821979 was deleted\n",
      "18:22:53.341 \u001b[33m[WARN]\u001b[0m taxid 283637 was merged into 3036909\n",
      "18:22:53.409 \u001b[33m[WARN]\u001b[0m taxid 2843217 was merged into 32224\n",
      "18:22:53.454 \u001b[33m[WARN]\u001b[0m taxid 285237 was merged into 3149680\n",
      "18:22:53.458 \u001b[33m[WARN]\u001b[0m taxid 2853430 was merged into 2854036\n",
      "18:22:53.472 \u001b[33m[WARN]\u001b[0m taxid 2855502 was merged into 2872311\n",
      "18:22:53.473 \u001b[33m[WARN]\u001b[0m taxid 285567 was merged into 1942\n",
      "18:22:53.475 \u001b[33m[WARN]\u001b[0m taxid 285570 was merged into 67282\n",
      "18:22:53.475 \u001b[33m[WARN]\u001b[0m taxid 285571 was merged into 67282\n",
      "18:22:53.884 \u001b[33m[WARN]\u001b[0m taxid 2888782 was merged into 3336977\n",
      "18:22:53.884 \u001b[33m[WARN]\u001b[0m taxid 2888783 was merged into 3336978\n",
      "18:22:53.884 \u001b[33m[WARN]\u001b[0m taxid 2888784 was merged into 3336975\n",
      "18:22:53.884 \u001b[33m[WARN]\u001b[0m taxid 2888785 was merged into 3336975\n",
      "18:22:53.884 \u001b[33m[WARN]\u001b[0m taxid 2888786 was merged into 3336974\n",
      "18:22:53.884 \u001b[33m[WARN]\u001b[0m taxid 2888787 was merged into 3336974\n",
      "18:22:53.884 \u001b[33m[WARN]\u001b[0m taxid 2888788 was merged into 3336973\n",
      "18:22:53.884 \u001b[33m[WARN]\u001b[0m taxid 2888789 was merged into 3336973\n",
      "18:22:53.884 \u001b[33m[WARN]\u001b[0m taxid 2888781 was merged into 3336976\n",
      "18:22:53.929 \u001b[33m[WARN]\u001b[0m taxid 2898584 was merged into 2949090\n",
      "18:22:53.945 \u001b[33m[WARN]\u001b[0m taxid 2903535 was merged into 2905958\n",
      "18:22:53.995 \u001b[33m[WARN]\u001b[0m taxid 2916726 was merged into 2913504\n",
      "18:22:53.996 \u001b[33m[WARN]\u001b[0m taxid 2916757 was merged into 2779506\n",
      "18:22:54.003 \u001b[33m[WARN]\u001b[0m taxid 2917990 was merged into 2805427\n",
      "18:22:54.124 \u001b[33m[WARN]\u001b[0m taxid 2949646 was merged into 2949647\n",
      "18:22:54.131 \u001b[33m[WARN]\u001b[0m taxid 2954076 was merged into 3237298\n",
      "18:22:54.140 \u001b[33m[WARN]\u001b[0m taxid 2961891 was merged into 2961893\n",
      "18:22:54.140 \u001b[33m[WARN]\u001b[0m taxid 2961892 was merged into 2961893\n",
      "18:22:54.171 \u001b[33m[WARN]\u001b[0m taxid 2972456 was merged into 2710754\n",
      "18:22:54.173 \u001b[33m[WARN]\u001b[0m taxid 29728 was merged into 2982281\n",
      "18:22:54.179 \u001b[33m[WARN]\u001b[0m taxid 2975279 was merged into 2975278\n",
      "18:22:54.183 \u001b[33m[WARN]\u001b[0m taxid 2976769 was merged into 427589\n",
      "18:22:54.185 \u001b[33m[WARN]\u001b[0m taxid 2978352 was deleted\n",
      "18:22:54.188 \u001b[33m[WARN]\u001b[0m taxid 2979989 was merged into 2979990\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983733 was merged into 3041312\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983734 was merged into 3041312\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983735 was merged into 3041312\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983736 was merged into 3041312\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983752 was merged into 3041311\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983737 was merged into 3041312\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983738 was merged into 3041312\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983753 was merged into 3041311\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983739 was merged into 3041312\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983740 was merged into 3041312\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983754 was merged into 3041311\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983742 was merged into 3041312\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983741 was merged into 3041312\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983732 was merged into 3041312\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983755 was merged into 3041311\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983756 was merged into 3041311\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983743 was merged into 3041313\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983757 was merged into 3041311\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983744 was merged into 3041313\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983758 was merged into 3041311\n",
      "18:22:54.194 \u001b[33m[WARN]\u001b[0m taxid 2983745 was merged into 3041313\n",
      "18:22:54.195 \u001b[33m[WARN]\u001b[0m taxid 2983746 was merged into 3041313\n",
      "18:22:54.195 \u001b[33m[WARN]\u001b[0m taxid 2983759 was merged into 3041311\n",
      "18:22:54.195 \u001b[33m[WARN]\u001b[0m taxid 2983747 was merged into 3041313\n",
      "18:22:54.195 \u001b[33m[WARN]\u001b[0m taxid 2983748 was merged into 3210771\n",
      "18:22:54.195 \u001b[33m[WARN]\u001b[0m taxid 2983760 was merged into 3041311\n",
      "18:22:54.195 \u001b[33m[WARN]\u001b[0m taxid 2983749 was merged into 3210771\n",
      "18:22:54.195 \u001b[33m[WARN]\u001b[0m taxid 2983750 was merged into 3041311\n",
      "18:22:54.195 \u001b[33m[WARN]\u001b[0m taxid 2983751 was merged into 3041311\n",
      "18:22:54.221 \u001b[33m[WARN]\u001b[0m taxid 299123 was merged into 40157\n",
      "18:22:54.232 \u001b[33m[WARN]\u001b[0m taxid 2993317 was merged into 1604349\n",
      "18:22:54.247 \u001b[33m[WARN]\u001b[0m taxid 2999065 was merged into 2999064\n",
      "18:22:54.289 \u001b[33m[WARN]\u001b[0m taxid 3024863 was merged into 3024868\n",
      "18:22:54.319 \u001b[33m[WARN]\u001b[0m taxid 3038048 was merged into 3037989\n",
      "18:22:54.324 \u001b[33m[WARN]\u001b[0m taxid 304008 was merged into 3336989\n",
      "18:22:54.330 \u001b[33m[WARN]\u001b[0m taxid 3043294 was merged into 3043293\n",
      "18:22:54.330 \u001b[33m[WARN]\u001b[0m taxid 3043295 was merged into 3043296\n",
      "18:22:54.333 \u001b[33m[WARN]\u001b[0m taxid 304600 was merged into 3150902\n",
      "18:22:54.333 \u001b[33m[WARN]\u001b[0m taxid 304613 was merged into 3150147\n",
      "18:22:54.334 \u001b[33m[WARN]\u001b[0m taxid 304654 was merged into 3149571\n",
      "18:22:54.334 \u001b[33m[WARN]\u001b[0m taxid 304674 was merged into 472785\n",
      "18:22:54.334 \u001b[33m[WARN]\u001b[0m taxid 304678 was merged into 472824\n",
      "18:22:54.339 \u001b[33m[WARN]\u001b[0m taxid 3051547 was deleted\n",
      "18:22:54.339 \u001b[33m[WARN]\u001b[0m taxid 3051661 was merged into 3051663\n",
      "18:22:54.339 \u001b[33m[WARN]\u001b[0m taxid 3051662 was merged into 3051663\n",
      "18:22:54.339 \u001b[33m[WARN]\u001b[0m taxid 3051827 was merged into 3051663\n",
      "18:22:54.339 \u001b[33m[WARN]\u001b[0m taxid 3053374 was deleted\n",
      "18:22:54.339 \u001b[33m[WARN]\u001b[0m taxid 3053354 was merged into 3046670\n",
      "18:22:54.345 \u001b[33m[WARN]\u001b[0m taxid 3058360 was merged into 2971264\n",
      "18:22:54.345 \u001b[33m[WARN]\u001b[0m taxid 3058396 was merged into 3058393\n",
      "18:22:54.345 \u001b[33m[WARN]\u001b[0m taxid 3058397 was merged into 3058394\n",
      "18:22:54.345 \u001b[33m[WARN]\u001b[0m taxid 3058399 was merged into 3058398\n",
      "18:22:54.370 \u001b[33m[WARN]\u001b[0m taxid 3067308 was merged into 3067307\n",
      "18:22:54.391 \u001b[33m[WARN]\u001b[0m taxid 3081589 was merged into 3297698\n",
      "18:22:54.402 \u001b[33m[WARN]\u001b[0m taxid 3090834 was merged into 3090833\n",
      "18:22:54.402 \u001b[33m[WARN]\u001b[0m taxid 3090835 was merged into 3090833\n",
      "18:22:54.402 \u001b[33m[WARN]\u001b[0m taxid 3090836 was merged into 3090833\n",
      "18:22:54.402 \u001b[33m[WARN]\u001b[0m taxid 3090837 was merged into 3090833\n",
      "18:22:54.403 \u001b[33m[WARN]\u001b[0m taxid 3092586 was merged into 3092585\n",
      "18:22:54.406 \u001b[33m[WARN]\u001b[0m taxid 3097448 was merged into 3080056\n",
      "18:22:54.407 \u001b[33m[WARN]\u001b[0m taxid 3097449 was merged into 3080056\n",
      "18:22:54.407 \u001b[33m[WARN]\u001b[0m taxid 3097450 was merged into 3080056\n",
      "18:22:54.416 \u001b[33m[WARN]\u001b[0m taxid 3106010 was merged into 2202420\n",
      "18:22:54.416 \u001b[33m[WARN]\u001b[0m taxid 3106011 was merged into 2202420\n",
      "18:22:54.420 \u001b[33m[WARN]\u001b[0m taxid 3110287 was merged into 3110288\n",
      "18:22:54.420 \u001b[33m[WARN]\u001b[0m taxid 3110289 was merged into 3110288\n",
      "18:22:54.420 \u001b[33m[WARN]\u001b[0m taxid 3110290 was merged into 3110288\n",
      "18:22:54.425 \u001b[33m[WARN]\u001b[0m taxid 3115495 was merged into 2602105\n",
      "18:22:54.439 \u001b[33m[WARN]\u001b[0m taxid 3127119 was merged into 1763824\n",
      "18:22:54.454 \u001b[33m[WARN]\u001b[0m taxid 3142549 was merged into 3063535\n",
      "18:22:54.454 \u001b[33m[WARN]\u001b[0m taxid 3142550 was merged into 3063535\n",
      "18:22:54.454 \u001b[33m[WARN]\u001b[0m taxid 3142551 was merged into 3063538\n",
      "18:22:54.454 \u001b[33m[WARN]\u001b[0m taxid 3142552 was merged into 3063538\n",
      "18:22:54.466 \u001b[33m[WARN]\u001b[0m taxid 3157705 was merged into 392011\n",
      "18:22:54.466 \u001b[33m[WARN]\u001b[0m taxid 3157706 was merged into 392011\n",
      "18:22:54.484 \u001b[33m[WARN]\u001b[0m taxid 322609 was merged into 85044\n",
      "18:22:54.491 \u001b[33m[WARN]\u001b[0m taxid 324069 was merged into 2763338\n",
      "18:22:54.498 \u001b[33m[WARN]\u001b[0m taxid 328552 was merged into 2734912\n",
      "18:22:54.500 \u001b[33m[WARN]\u001b[0m taxid 329913 was merged into 278785\n",
      "18:22:54.502 \u001b[33m[WARN]\u001b[0m taxid 331009 was merged into 3150466\n",
      "18:22:54.522 \u001b[33m[WARN]\u001b[0m taxid 348592 was merged into 2547900\n",
      "18:22:54.523 \u001b[33m[WARN]\u001b[0m taxid 348819 was merged into 2744967\n",
      "18:22:54.523 \u001b[33m[WARN]\u001b[0m taxid 348820 was merged into 3150384\n",
      "18:22:54.536 \u001b[33m[WARN]\u001b[0m taxid 357742 was merged into 3149785\n",
      "18:22:54.547 \u001b[33m[WARN]\u001b[0m taxid 370325 was merged into 3150081\n",
      "18:22:54.555 \u001b[33m[WARN]\u001b[0m taxid 380366 was merged into 3149453\n",
      "18:22:54.556 \u001b[33m[WARN]\u001b[0m taxid 381198 was merged into 8845\n",
      "18:22:54.565 \u001b[33m[WARN]\u001b[0m taxid 385651 was merged into 3315559\n",
      "18:22:54.578 \u001b[33m[WARN]\u001b[0m taxid 390426 was merged into 2969111\n",
      "18:22:54.590 \u001b[33m[WARN]\u001b[0m taxid 393726 was merged into 3149873\n",
      "18:22:54.606 \u001b[33m[WARN]\u001b[0m taxid 401876 was merged into 741065\n",
      "18:22:54.622 \u001b[33m[WARN]\u001b[0m taxid 4111 was merged into 223891\n",
      "18:22:54.661 \u001b[33m[WARN]\u001b[0m taxid 424880 was merged into 3150040\n",
      "18:22:54.723 \u001b[33m[WARN]\u001b[0m taxid 44964 was merged into 44972\n",
      "18:22:54.740 \u001b[33m[WARN]\u001b[0m taxid 460221 was merged into 3151013\n",
      "18:22:54.740 \u001b[33m[WARN]\u001b[0m taxid 460227 was merged into 3149827\n",
      "18:22:54.741 \u001b[33m[WARN]\u001b[0m taxid 461224 was merged into 3150382\n",
      "18:22:54.741 \u001b[33m[WARN]\u001b[0m taxid 461227 was merged into 3150339\n",
      "18:22:54.780 \u001b[33m[WARN]\u001b[0m taxid 4790 was merged into 4792\n",
      "18:22:54.785 \u001b[33m[WARN]\u001b[0m taxid 482479 was merged into 1120086\n",
      "18:22:54.786 \u001b[33m[WARN]\u001b[0m taxid 482926 was merged into 2934751\n",
      "18:22:54.807 \u001b[33m[WARN]\u001b[0m taxid 495235 was merged into 3149650\n",
      "18:22:54.882 \u001b[33m[WARN]\u001b[0m taxid 520176 was merged into 3151029\n",
      "18:22:54.982 \u001b[33m[WARN]\u001b[0m taxid 54492 was merged into 3150717\n",
      "18:22:54.982 \u001b[33m[WARN]\u001b[0m taxid 54493 was merged into 3150718\n",
      "18:22:55.064 \u001b[33m[WARN]\u001b[0m taxid 558553 was merged into 2784310\n",
      "18:22:55.115 \u001b[33m[WARN]\u001b[0m taxid 572152 was merged into 3149918\n",
      "18:22:55.158 \u001b[33m[WARN]\u001b[0m taxid 58251 was merged into 227531\n",
      "18:22:55.285 \u001b[33m[WARN]\u001b[0m taxid 63296 was merged into 63294\n",
      "18:22:55.315 \u001b[33m[WARN]\u001b[0m taxid 644345 was merged into 3273801\n",
      "18:22:55.334 \u001b[33m[WARN]\u001b[0m taxid 64815 was merged into 3149790\n",
      "18:22:55.355 \u001b[33m[WARN]\u001b[0m taxid 65497 was merged into 2893586\n",
      "18:22:55.384 \u001b[33m[WARN]\u001b[0m taxid 66372 was merged into 68174\n",
      "18:22:55.400 \u001b[33m[WARN]\u001b[0m taxid 66873 was merged into 116188\n",
      "18:22:55.401 \u001b[33m[WARN]\u001b[0m taxid 66889 was merged into 28893\n",
      "18:22:55.401 \u001b[33m[WARN]\u001b[0m taxid 66898 was merged into 66895\n",
      "18:22:55.408 \u001b[33m[WARN]\u001b[0m taxid 67328 was merged into 67384\n",
      "18:22:55.408 \u001b[33m[WARN]\u001b[0m taxid 67349 was merged into 1894\n",
      "18:22:55.427 \u001b[33m[WARN]\u001b[0m taxid 68178 was merged into 1894\n",
      "18:22:55.427 \u001b[33m[WARN]\u001b[0m taxid 68189 was merged into 68174\n",
      "18:22:55.428 \u001b[33m[WARN]\u001b[0m taxid 68202 was merged into 1919\n",
      "18:22:55.428 \u001b[33m[WARN]\u001b[0m taxid 68210 was merged into 47760\n",
      "18:22:55.440 \u001b[33m[WARN]\u001b[0m taxid 68570 was merged into 1971\n",
      "18:22:55.459 \u001b[33m[WARN]\u001b[0m taxid 690698 was merged into 3144495\n",
      "18:22:55.469 \u001b[33m[WARN]\u001b[0m taxid 69262 was merged into 2857398\n",
      "18:22:55.510 \u001b[33m[WARN]\u001b[0m taxid 70791 was merged into 2747968\n",
      "18:22:55.512 \u001b[33m[WARN]\u001b[0m taxid 70892 was merged into 7757\n",
      "18:22:55.548 \u001b[33m[WARN]\u001b[0m taxid 743428 was merged into 3085985\n",
      "18:22:55.553 \u001b[33m[WARN]\u001b[0m taxid 746493 was merged into 3030045\n",
      "18:22:55.589 \u001b[33m[WARN]\u001b[0m taxid 758445 was merged into 3150094\n",
      "18:22:55.629 \u001b[33m[WARN]\u001b[0m taxid 81037 was merged into 77608\n",
      "18:22:55.643 \u001b[33m[WARN]\u001b[0m taxid 83383 was merged into 67370\n",
      "18:22:55.681 \u001b[33m[WARN]\u001b[0m taxid 868594 was merged into 2923069\n",
      "18:22:55.717 \u001b[33m[WARN]\u001b[0m taxid 886884 was merged into 886883\n",
      "18:22:55.738 \u001b[33m[WARN]\u001b[0m taxid 91773 was merged into 2744969\n",
      "18:22:55.746 \u001b[33m[WARN]\u001b[0m taxid 929498 was merged into 3074752\n",
      "18:22:55.772 \u001b[33m[WARN]\u001b[0m taxid 947500 was merged into 2483980\n",
      "18:22:55.772 \u001b[33m[WARN]\u001b[0m taxid 947504 was merged into 2712986\n",
      "18:22:55.781 \u001b[33m[WARN]\u001b[0m taxid 979572 was merged into 3150076\n",
      "18:22:55.804 \u001b[33m[WARN]\u001b[0m taxid 990144 was merged into 588\n"
     ]
    }
   ],
   "source": [
    "# %%bash\n",
    "# gunzip prot.accession2taxid.FULL.gz\n",
    "\n",
    "# awk '{print $2}' prot.accession2taxid.FULL \\\n",
    "#     | sort -u \\\n",
    "#     | taxonkit --data-dir ncbi-taxdump lineage \\\n",
    "#     | rg \"\\tViruses;\" \\\n",
    "#     | awk '{print $1}' \\\n",
    "#     > virus_taxid.list\n",
    "# csvtk grep --quiet -t -f 2 -P virus_taxid.list prot.accession2taxid.FULL > virus.accession2taxid\n",
    "# # rm prot.accession2taxid.FULL # why antonio why I thought it crashed and then it deleted it now I need to redownload this wasteful legacy filE??"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Find the ICTV-compliant proteins and write a new table with the ICTV taxids\n",
    "Execute the `get_ictv_taxids.py` (moved into the code block below) script to create a `accession2taxid` file with ICTV taxids."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "import taxopy\n",
    "\n",
    "# Read the official ICTV names\n",
    "with open(\"ictv.names.txt\") as fin:\n",
    "    ictv_name_set = {i.strip() for i in fin.readlines()}\n",
    "\n",
    "# Create the Taxopy NCBI and ICTV databases\n",
    "ncbi_taxdb = taxopy.TaxDb(\n",
    "    nodes_dmp=\"ncbi-taxdump/nodes.dmp\",\n",
    "    names_dmp=\"ncbi-taxdump/names.dmp\",\n",
    "    merged_dmp=\"ncbi-taxdump/merged.dmp\",\n",
    "    keep_files=True,\n",
    ")\n",
    "ictv_taxdb = taxopy.TaxDb(\n",
    "    nodes_dmp=\"ictv-taxdump/nodes.dmp\",\n",
    "    names_dmp=\"ictv-taxdump/names.dmp\",\n",
    "    keep_files=True,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Replace non-ICTV taxids\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"virus.accession2taxid.tsv\") as fin:\n",
    " with open(\"virus.accession2taxid.ictv\", \"w\") as fout:\n",
    "    for line in fin:\n",
    "        acc, taxid = line.split(\"\\t\")\n",
    "        ncbi_taxon = taxopy.Taxon(int(taxid), ncbi_taxdb)\n",
    "        for j in ncbi_taxon.name_lineage:\n",
    "            if j in ictv_name_set:\n",
    "                ictv_taxid = taxopy.taxid_from_name(j, ictv_taxdb)\n",
    "                ictv_taxon = taxopy.Taxon(ictv_taxid[0], ictv_taxdb)\n",
    "                fout.write(f\"{acc}\\t{ictv_taxon.taxid}\\n\")\n",
    "                # break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Concatenate the virus proteins associated with ICTV viruses:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# # Create a list containing the accessions of the proteins of ICTV viruses\n",
    "cut -f 1 virus.accession2taxid.ictv > virus.accession.txt\n",
    "\n",
    "# # Filter the NR proteins to keep the proteins encoded by ICTV viruses\n",
    "# seqkit grep -j 4 -f virus.accession.txt nr.gz | seqkit seq -i -w 0 -o nr.virus.faa.gz\n",
    "seqkit grep -j 4 -f virus.accession.txt concatenated_proteins.fasta | seqkit seq -i -w 0 -o nr.virus.faa.gz\n",
    "\n",
    "# # Download and filter NR proteins\n",
    "# aria2c --check-certificate=false -x 4 \"https://ftp.ncbi.nlm.nih.gov/blast/db/FASTA/nr.gz\"\n",
    "# rm nr.gz # why antonio "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There will be proteins in `virus.accession2taxid.ictv` that are not in NR. So we will keep only the proteins that are present in the filtered NR FASTA file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO]\u001b[0m processed records: 8192\n",
      "[INFO]\u001b[0m processed records: 16384\n",
      "[INFO]\u001b[0m processed records: 24576\n",
      "[INFO]\u001b[0m processed records: 32768\n",
      "[INFO]\u001b[0m processed records: 40960\n",
      "[INFO]\u001b[0m processed records: 49152\n",
      "[INFO]\u001b[0m processed records: 57344\n",
      "[INFO]\u001b[0m processed records: 65536\n",
      "[INFO]\u001b[0m processed records: 73728\n",
      "[INFO]\u001b[0m processed records: 81920\n",
      "[INFO]\u001b[0m processed records: 90112\n",
      "[INFO]\u001b[0m processed records: 98304\n",
      "[INFO]\u001b[0m processed records: 106496\n",
      "[INFO]\u001b[0m processed records: 114688\n",
      "[INFO]\u001b[0m processed records: 122880\n",
      "[INFO]\u001b[0m processed records: 131072\n",
      "[INFO]\u001b[0m processed records: 139264\n",
      "[INFO]\u001b[0m processed records: 147456\n",
      "[INFO]\u001b[0m processed records: 155648\n",
      "[INFO]\u001b[0m processed records: 163840\n",
      "[INFO]\u001b[0m processed records: 172032\n",
      "[INFO]\u001b[0m processed records: 180224\n",
      "[INFO]\u001b[0m processed records: 188416\n",
      "[INFO]\u001b[0m processed records: 196608\n",
      "[INFO]\u001b[0m processed records: 204800\n",
      "[INFO]\u001b[0m processed records: 212992\n",
      "[INFO]\u001b[0m processed records: 221184\n",
      "[INFO]\u001b[0m processed records: 229376\n",
      "[INFO]\u001b[0m processed records: 237568\n",
      "[INFO]\u001b[0m processed records: 245760\n",
      "[INFO]\u001b[0m processed records: 253952\n",
      "[INFO]\u001b[0m processed records: 262144\n",
      "[INFO]\u001b[0m processed records: 270336\n",
      "[INFO]\u001b[0m processed records: 278528\n",
      "[INFO]\u001b[0m processed records: 286720\n",
      "[INFO]\u001b[0m processed records: 294912\n",
      "[INFO]\u001b[0m processed records: 303104\n",
      "[INFO]\u001b[0m processed records: 311296\n",
      "[INFO]\u001b[0m processed records: 319488\n",
      "[INFO]\u001b[0m processed records: 327680\n",
      "[INFO]\u001b[0m processed records: 335872\n",
      "[INFO]\u001b[0m processed records: 344064\n",
      "[INFO]\u001b[0m processed records: 352256\n",
      "[INFO]\u001b[0m processed records: 360448\n",
      "[INFO]\u001b[0m processed records: 368640\n",
      "[INFO]\u001b[0m processed records: 376832\n",
      "[INFO]\u001b[0m processed records: 385024\n",
      "[INFO]\u001b[0m processed records: 393216\n",
      "[INFO]\u001b[0m processed records: 401408\n",
      "[INFO]\u001b[0m processed records: 409600\n",
      "[INFO]\u001b[0m processed records: 417792\n",
      "[INFO]\u001b[0m processed records: 425984\n",
      "[INFO]\u001b[0m processed records: 434176\n",
      "[INFO]\u001b[0m processed records: 442368\n",
      "[INFO]\u001b[0m processed records: 450560\n",
      "[INFO]\u001b[0m processed records: 458752\n",
      "[INFO]\u001b[0m processed records: 466944\n",
      "[INFO]\u001b[0m processed records: 475136\n",
      "[INFO]\u001b[0m processed records: 483328\n",
      "[INFO]\u001b[0m processed records: 491520\n",
      "[INFO]\u001b[0m processed records: 499712\n",
      "[INFO]\u001b[0m processed records: 507904\n",
      "[INFO]\u001b[0m processed records: 516096\n",
      "[INFO]\u001b[0m processed records: 524288\n",
      "[INFO]\u001b[0m processed records: 532480\n",
      "[INFO]\u001b[0m processed records: 540672\n",
      "[INFO]\u001b[0m processed records: 548864\n",
      "[INFO]\u001b[0m processed records: 557056\n",
      "[INFO]\u001b[0m processed records: 565248\n",
      "[INFO]\u001b[0m processed records: 573440\n",
      "[INFO]\u001b[0m processed records: 581632\n",
      "[INFO]\u001b[0m processed records: 589824\n",
      "[INFO]\u001b[0m processed records: 598016\n",
      "[INFO]\u001b[0m processed records: 606208\n",
      "[INFO]\u001b[0m processed records: 614400\n",
      "[INFO]\u001b[0m processed records: 622592\n",
      "[INFO]\u001b[0m processed records: 630784\n",
      "[INFO]\u001b[0m processed records: 638976\n",
      "[INFO]\u001b[0m processed records: 647168\n",
      "[INFO]\u001b[0m processed records: 655360\n",
      "[INFO]\u001b[0m processed records: 663552\n",
      "[INFO]\u001b[0m processed records: 671744\n",
      "[INFO]\u001b[0m processed records: 679936\n",
      "[INFO]\u001b[0m processed records: 688128\n",
      "[INFO]\u001b[0m processed records: 696320\n",
      "[INFO]\u001b[0m processed records: 704512\n",
      "[INFO]\u001b[0m processed records: 712704\n",
      "[INFO]\u001b[0m processed records: 720896\n",
      "[INFO]\u001b[0m processed records: 729088\n",
      "[INFO]\u001b[0m processed records: 737280\n",
      "[INFO]\u001b[0m processed records: 745472\n",
      "[INFO]\u001b[0m processed records: 753664\n",
      "[INFO]\u001b[0m processed records: 761856\n",
      "[INFO]\u001b[0m processed records: 770048\n",
      "[INFO]\u001b[0m processed records: 778240\n",
      "[INFO]\u001b[0m processed records: 786432\n",
      "[INFO]\u001b[0m processed records: 794624\n",
      "[INFO]\u001b[0m processed records: 802816\n",
      "[INFO]\u001b[0m processed records: 811008\n",
      "[INFO]\u001b[0m processed records: 819200\n",
      "[INFO]\u001b[0m processed records: 827392\n",
      "[INFO]\u001b[0m processed records: 835584\n",
      "[INFO]\u001b[0m processed records: 843776\n",
      "[INFO]\u001b[0m processed records: 851968\n",
      "[INFO]\u001b[0m processed records: 860160\n",
      "[INFO]\u001b[0m processed records: 868352\n",
      "[INFO]\u001b[0m processed records: 876544\n",
      "[INFO]\u001b[0m processed records: 884736\n",
      "[INFO]\u001b[0m processed records: 892928\n",
      "[INFO]\u001b[0m processed records: 901120\n",
      "[INFO]\u001b[0m processed records: 909312\n",
      "[INFO]\u001b[0m processed records: 917504\n",
      "[INFO]\u001b[0m processed records: 925696\n",
      "[INFO]\u001b[0m processed records: 933888\n",
      "[INFO]\u001b[0m processed records: 942080\n",
      "[INFO]\u001b[0m processed records: 950272\n",
      "[INFO]\u001b[0m processed records: 958464\n",
      "[INFO]\u001b[0m processed records: 966656\n",
      "[INFO]\u001b[0m processed records: 974848\n",
      "[INFO]\u001b[0m processed records: 983040\n",
      "[INFO]\u001b[0m processed records: 991232\n",
      "[INFO]\u001b[0m processed records: 999424\n",
      "[INFO]\u001b[0m processed records: 1007616\n",
      "[INFO]\u001b[0m processed records: 1015808\n",
      "[INFO]\u001b[0m processed records: 1024000\n",
      "[INFO]\u001b[0m processed records: 1032192\n",
      "[INFO]\u001b[0m processed records: 1040384\n",
      "[INFO]\u001b[0m processed records: 1048576\n",
      "[INFO]\u001b[0m processed records: 1056768\n",
      "[INFO]\u001b[0m processed records: 1064960\n",
      "[INFO]\u001b[0m processed records: 1073152\n",
      "[INFO]\u001b[0m processed records: 1081344\n",
      "[INFO]\u001b[0m processed records: 1089536\n",
      "[INFO]\u001b[0m processed records: 1097728\n",
      "[INFO]\u001b[0m processed records: 1105920\n",
      "[INFO]\u001b[0m processed records: 1114112\n",
      "[INFO]\u001b[0m processed records: 1122304\n",
      "[INFO]\u001b[0m processed records: 1130496\n",
      "[INFO]\u001b[0m processed records: 1138688\n",
      "[INFO]\u001b[0m processed records: 1146880\n",
      "[INFO]\u001b[0m processed records: 1155072\n",
      "[INFO]\u001b[0m processed records: 1163264\n",
      "[INFO]\u001b[0m processed records: 1171456\n",
      "[INFO]\u001b[0m processed records: 1179648\n",
      "[INFO]\u001b[0m processed records: 1187840\n",
      "[INFO]\u001b[0m processed records: 1196032\n",
      "[INFO]\u001b[0m processed records: 1204224\n",
      "[INFO]\u001b[0m processed records: 1212416\n",
      "[INFO]\u001b[0m processed records: 1220608\n",
      "[INFO]\u001b[0m processed records: 1228800\n",
      "[INFO]\u001b[0m processed records: 1236992\n",
      "[INFO]\u001b[0m processed records: 1245184\n",
      "[INFO]\u001b[0m processed records: 1253376\n",
      "[INFO]\u001b[0m processed records: 1261568\n",
      "[INFO]\u001b[0m processed records: 1269760\n",
      "[INFO]\u001b[0m processed records: 1277952\n",
      "[INFO]\u001b[0m processed records: 1286144\n",
      "[INFO]\u001b[0m processed records: 1294336\n",
      "[INFO]\u001b[0m processed records: 1302528\n",
      "[INFO]\u001b[0m processed records: 1310720\n",
      "[INFO]\u001b[0m processed records: 1318912\n",
      "[INFO]\u001b[0m processed records: 1327104\n",
      "[INFO]\u001b[0m processed records: 1335296\n",
      "[INFO]\u001b[0m processed records: 1343488\n",
      "[INFO]\u001b[0m processed records: 1351680\n",
      "[INFO]\u001b[0m processed records: 1359872\n",
      "[INFO]\u001b[0m processed records: 1368064\n",
      "[INFO]\u001b[0m processed records: 1376256\n",
      "[INFO]\u001b[0m processed records: 1384448\n",
      "[INFO]\u001b[0m processed records: 1392640\n",
      "[INFO]\u001b[0m processed records: 1400832\n",
      "[INFO]\u001b[0m processed records: 1409024\n",
      "[INFO]\u001b[0m processed records: 1417216\n",
      "[INFO]\u001b[0m processed records: 1425408\n",
      "[INFO]\u001b[0m processed records: 1433600\n",
      "[INFO]\u001b[0m processed records: 1441792\n",
      "[INFO]\u001b[0m processed records: 1449984\n",
      "[INFO]\u001b[0m processed records: 1458176\n",
      "[INFO]\u001b[0m processed records: 1466368\n",
      "[INFO]\u001b[0m processed records: 1474560\n",
      "[INFO]\u001b[0m processed records: 1482752\n",
      "[INFO]\u001b[0m processed records: 1490944\n",
      "[INFO]\u001b[0m processed records: 1499136\n",
      "[INFO]\u001b[0m processed records: 1507328\n",
      "[INFO]\u001b[0m processed records: 1515520\n",
      "[INFO]\u001b[0m processed records: 1523712\n",
      "[INFO]\u001b[0m processed records: 1531904\n",
      "[INFO]\u001b[0m processed records: 1540096\n",
      "[INFO]\u001b[0m processed records: 1548288\n",
      "[INFO]\u001b[0m processed records: 1556480\n",
      "[INFO]\u001b[0m processed records: 1564672\n",
      "[INFO]\u001b[0m processed records: 1572864\n",
      "[INFO]\u001b[0m processed records: 1581056\n",
      "[INFO]\u001b[0m processed records: 1589248\n",
      "[INFO]\u001b[0m processed records: 1597440\n",
      "[INFO]\u001b[0m processed records: 1605632\n",
      "[INFO]\u001b[0m processed records: 1613824\n",
      "[INFO]\u001b[0m processed records: 1622016\n",
      "[INFO]\u001b[0m processed records: 1630208\n",
      "[INFO]\u001b[0m processed records: 1638400\n",
      "[INFO]\u001b[0m processed records: 1646592\n",
      "[INFO]\u001b[0m processed records: 1654784\n",
      "[INFO]\u001b[0m processed records: 1662976\n",
      "[INFO]\u001b[0m processed records: 1671168\n",
      "[INFO]\u001b[0m processed records: 1679360\n",
      "[INFO]\u001b[0m processed records: 1687552\n",
      "[INFO]\u001b[0m processed records: 1695744\n",
      "[INFO]\u001b[0m processed records: 1703936\n",
      "[INFO]\u001b[0m processed records: 1712128\n",
      "[INFO]\u001b[0m processed records: 1720320\n",
      "[INFO]\u001b[0m processed records: 1728512\n",
      "[INFO]\u001b[0m processed records: 1736704\n",
      "[INFO]\u001b[0m processed records: 1744896\n",
      "[INFO]\u001b[0m processed records: 1753088\n",
      "[INFO]\u001b[0m processed records: 1761280\n",
      "[INFO]\u001b[0m processed records: 1769472\n",
      "[INFO]\u001b[0m processed records: 1777664\n",
      "[INFO]\u001b[0m processed records: 1785856\n",
      "[INFO]\u001b[0m processed records: 1794048\n",
      "[INFO]\u001b[0m processed records: 1802240\n",
      "[INFO]\u001b[0m processed records: 1810432\n",
      "[INFO]\u001b[0m processed records: 1818624\n",
      "[INFO]\u001b[0m processed records: 1826816\n",
      "[INFO]\u001b[0m processed records: 1835008\n",
      "[INFO]\u001b[0m processed records: 1843200\n",
      "[INFO]\u001b[0m processed records: 1851392\n",
      "[INFO]\u001b[0m processed records: 1859584\n",
      "[INFO]\u001b[0m processed records: 1867776\n",
      "[INFO]\u001b[0m processed records: 1875968\n",
      "[INFO]\u001b[0m processed records: 1884160\n",
      "[INFO]\u001b[0m processed records: 1892352\n",
      "[INFO]\u001b[0m processed records: 1900544\n",
      "[INFO]\u001b[0m processed records: 1908736\n",
      "[INFO]\u001b[0m processed records: 1916928\n",
      "[INFO]\u001b[0m processed records: 1925120\n",
      "[INFO]\u001b[0m processed records: 1933312\n",
      "[INFO]\u001b[0m processed records: 1941504\n",
      "[INFO]\u001b[0m processed records: 1949696\n",
      "[INFO]\u001b[0m processed records: 1957888\n",
      "[INFO]\u001b[0m processed records: 1966080\n",
      "[INFO]\u001b[0m processed records: 1974272\n",
      "[INFO]\u001b[0m processed records: 1982464\n",
      "[INFO]\u001b[0m processed records: 1990656\n",
      "[INFO]\u001b[0m processed records: 1998848\n",
      "[INFO]\u001b[0m processed records: 2007040\n",
      "[INFO]\u001b[0m processed records: 2015232\n",
      "[INFO]\u001b[0m processed records: 2023424\n",
      "[INFO]\u001b[0m processed records: 2031616\n",
      "[INFO]\u001b[0m processed records: 2039808\n",
      "[INFO]\u001b[0m processed records: 2048000\n",
      "[INFO]\u001b[0m processed records: 2056192\n",
      "[INFO]\u001b[0m processed records: 2064384\n",
      "[INFO]\u001b[0m processed records: 2072576\n",
      "[INFO]\u001b[0m processed records: 2080768\n",
      "[INFO]\u001b[0m processed records: 2088960\n",
      "[INFO]\u001b[0m processed records: 2097152\n",
      "[INFO]\u001b[0m processed records: 2105344\n",
      "[INFO]\u001b[0m processed records: 2113536\n",
      "[INFO]\u001b[0m processed records: 2121728\n",
      "[INFO]\u001b[0m processed records: 2129920\n",
      "[INFO]\u001b[0m processed records: 2138112\n",
      "[INFO]\u001b[0m processed records: 2146304\n",
      "[INFO]\u001b[0m processed records: 2154496\n",
      "[INFO]\u001b[0m processed records: 2162688\n",
      "[INFO]\u001b[0m processed records: 2170880\n",
      "[INFO]\u001b[0m processed records: 2179072\n",
      "[INFO]\u001b[0m processed records: 2187264\n",
      "[INFO]\u001b[0m processed records: 2195456\n",
      "[INFO]\u001b[0m processed records: 2203648\n",
      "[INFO]\u001b[0m processed records: 2211840\n",
      "[INFO]\u001b[0m processed records: 2220032\n",
      "[INFO]\u001b[0m processed records: 2228224\n",
      "[INFO]\u001b[0m processed records: 2236416\n",
      "[INFO]\u001b[0m processed records: 2244608\n",
      "[INFO]\u001b[0m processed records: 2252800\n",
      "[INFO]\u001b[0m processed records: 2260992\n",
      "[INFO]\u001b[0m processed records: 2269184\n",
      "[INFO]\u001b[0m processed records: 2277376\n",
      "[INFO]\u001b[0m processed records: 2285568\n",
      "[INFO]\u001b[0m processed records: 2293760\n",
      "[INFO]\u001b[0m processed records: 2301952\n",
      "[INFO]\u001b[0m processed records: 2310144\n",
      "[INFO]\u001b[0m processed records: 2318336\n",
      "[INFO]\u001b[0m processed records: 2326528\n",
      "[INFO]\u001b[0m processed records: 2334720\n",
      "[INFO]\u001b[0m processed records: 2342912\n",
      "[INFO]\u001b[0m processed records: 2351104\n",
      "[INFO]\u001b[0m processed records: 2359296\n",
      "[INFO]\u001b[0m processed records: 2367488\n",
      "[INFO]\u001b[0m processed records: 2375680\n",
      "[INFO]\u001b[0m processed records: 2383872\n",
      "[INFO]\u001b[0m processed records: 2392064\n",
      "[INFO]\u001b[0m processed records: 2400256\n",
      "[INFO]\u001b[0m processed records: 2408448\n",
      "[INFO]\u001b[0m processed records: 2416640\n",
      "[INFO]\u001b[0m processed records: 2424832\n",
      "[INFO]\u001b[0m processed records: 2433024\n",
      "[INFO]\u001b[0m processed records: 2441216\n",
      "[INFO]\u001b[0m processed records: 2449408\n",
      "[INFO]\u001b[0m processed records: 2457600\n",
      "[INFO]\u001b[0m processed records: 2465792\n",
      "[INFO]\u001b[0m processed records: 2473984\n",
      "[INFO]\u001b[0m processed records: 2482176\n",
      "[INFO]\u001b[0m processed records: 2490368\n",
      "[INFO]\u001b[0m processed records: 2498560\n",
      "[INFO]\u001b[0m processed records: 2506752\n",
      "[INFO]\u001b[0m processed records: 2514944\n",
      "[INFO]\u001b[0m processed records: 2523136\n",
      "[INFO]\u001b[0m processed records: 2531328\n",
      "[INFO]\u001b[0m processed records: 2539520\n",
      "[INFO]\u001b[0m processed records: 2547712\n",
      "[INFO]\u001b[0m processed records: 2555904\n",
      "[INFO]\u001b[0m processed records: 2564096\n",
      "[INFO]\u001b[0m processed records: 2572288\n",
      "[INFO]\u001b[0m processed records: 2580480\n",
      "[INFO]\u001b[0m processed records: 2588672\n",
      "[INFO]\u001b[0m processed records: 2596864\n",
      "[INFO]\u001b[0m processed records: 2605056\n",
      "[INFO]\u001b[0m processed records: 2613248\n",
      "[INFO]\u001b[0m processed records: 2621440\n",
      "[INFO]\u001b[0m processed records: 2629632\n",
      "[INFO]\u001b[0m processed records: 2637824\n",
      "[INFO]\u001b[0m processed records: 2646016\n",
      "[INFO]\u001b[0m processed records: 2654208\n",
      "[INFO]\u001b[0m processed records: 2662400\n",
      "[INFO]\u001b[0m processed records: 2670592\n",
      "[INFO]\u001b[0m processed records: 2678784\n",
      "[INFO]\u001b[0m processed records: 2686976\n",
      "[INFO]\u001b[0m processed records: 2695168\n",
      "[INFO]\u001b[0m processed records: 2703360\n",
      "[INFO]\u001b[0m processed records: 2711552\n",
      "[INFO]\u001b[0m processed records: 2719744\n",
      "[INFO]\u001b[0m processed records: 2727936\n",
      "[INFO]\u001b[0m processed records: 2736128\n",
      "[INFO]\u001b[0m processed records: 2744320\n",
      "[INFO]\u001b[0m processed records: 2752512\n",
      "[INFO]\u001b[0m processed records: 2760704\n",
      "[INFO]\u001b[0m processed records: 2768896\n",
      "[INFO]\u001b[0m processed records: 2777088\n",
      "[INFO]\u001b[0m processed records: 2785280\n",
      "[INFO]\u001b[0m processed records: 2793472\n",
      "[INFO]\u001b[0m processed records: 2801664\n",
      "[INFO]\u001b[0m processed records: 2809856\n",
      "[INFO]\u001b[0m processed records: 2818048\n",
      "[INFO]\u001b[0m processed records: 2826240\n",
      "[INFO]\u001b[0m processed records: 2834432\n",
      "[INFO]\u001b[0m processed records: 2842624\n",
      "[INFO]\u001b[0m processed records: 2850816\n",
      "[INFO]\u001b[0m processed records: 2859008\n",
      "[INFO]\u001b[0m processed records: 2867200\n",
      "[INFO]\u001b[0m processed records: 2875392\n",
      "[INFO]\u001b[0m processed records: 2883584\n",
      "[INFO]\u001b[0m processed records: 2891776\n",
      "[INFO]\u001b[0m processed records: 2899968\n",
      "[INFO]\u001b[0m processed records: 2908160\n",
      "[INFO]\u001b[0m processed records: 2916352\n",
      "[INFO]\u001b[0m processed records: 2924544\n",
      "[INFO]\u001b[0m processed records: 2932736\n",
      "[INFO]\u001b[0m processed records: 2940928\n",
      "[INFO]\u001b[0m processed records: 2949120\n",
      "[INFO]\u001b[0m processed records: 2957312\n",
      "[INFO]\u001b[0m processed records: 2965504\n",
      "[INFO]\u001b[0m processed records: 2973696\n",
      "[INFO]\u001b[0m processed records: 2981888\n",
      "[INFO]\u001b[0m processed records: 2990080\n",
      "[INFO]\u001b[0m processed records: 2998272\n",
      "[INFO]\u001b[0m processed records: 3006464\n",
      "[INFO]\u001b[0m processed records: 3014656\n",
      "[INFO]\u001b[0m processed records: 3022848\n",
      "[INFO]\u001b[0m processed records: 3031040\n",
      "[INFO]\u001b[0m processed records: 3039232\n",
      "[INFO]\u001b[0m processed records: 3047424\n",
      "[INFO]\u001b[0m processed records: 3055616\n",
      "[INFO]\u001b[0m processed records: 3063808\n",
      "[INFO]\u001b[0m processed records: 3072000\n",
      "[INFO]\u001b[0m processed records: 3080192\n",
      "[INFO]\u001b[0m processed records: 3088384\n",
      "[INFO]\u001b[0m processed records: 3096576\n",
      "[INFO]\u001b[0m processed records: 3104768\n",
      "[INFO]\u001b[0m processed records: 3112960\n",
      "[INFO]\u001b[0m processed records: 3121152\n",
      "[INFO]\u001b[0m processed records: 3129344\n",
      "[INFO]\u001b[0m processed records: 3137536\n",
      "[INFO]\u001b[0m processed records: 3145728\n",
      "[INFO]\u001b[0m processed records: 3153920\n",
      "[INFO]\u001b[0m processed records: 3162112\n",
      "[INFO]\u001b[0m processed records: 3170304\n",
      "[INFO]\u001b[0m processed records: 3178496\n",
      "[INFO]\u001b[0m processed records: 3186688\n",
      "[INFO]\u001b[0m processed records: 3194880\n",
      "[INFO]\u001b[0m processed records: 3203072\n",
      "[INFO]\u001b[0m processed records: 3211264\n",
      "[INFO]\u001b[0m processed records: 3219456\n",
      "[INFO]\u001b[0m processed records: 3227648\n",
      "[INFO]\u001b[0m processed records: 3235840\n",
      "[INFO]\u001b[0m processed records: 3244032\n",
      "[INFO]\u001b[0m processed records: 3252224\n",
      "[INFO]\u001b[0m processed records: 3260416\n",
      "[INFO]\u001b[0m processed records: 3268608\n",
      "[INFO]\u001b[0m processed records: 3276800\n",
      "[INFO]\u001b[0m processed records: 3284992\n",
      "[INFO]\u001b[0m processed records: 3293184\n",
      "[INFO]\u001b[0m processed records: 3301376\n",
      "[INFO]\u001b[0m processed records: 3309568\n",
      "[INFO]\u001b[0m processed records: 3317760\n",
      "[INFO]\u001b[0m processed records: 3325952\n",
      "[INFO]\u001b[0m processed records: 3334144\n",
      "[INFO]\u001b[0m processed records: 3342336\n",
      "[INFO]\u001b[0m processed records: 3350528\n",
      "[INFO]\u001b[0m processed records: 3358720\n",
      "[INFO]\u001b[0m processed records: 3366912\n",
      "[INFO]\u001b[0m processed records: 3375104\n",
      "[INFO]\u001b[0m processed records: 3383296\n",
      "[INFO]\u001b[0m processed records: 3391488\n",
      "[INFO]\u001b[0m processed records: 3399680\n",
      "[INFO]\u001b[0m processed records: 3407872\n",
      "[INFO]\u001b[0m processed records: 3416064\n",
      "[INFO]\u001b[0m processed records: 3424256\n",
      "[INFO]\u001b[0m processed records: 3432448\n",
      "[INFO]\u001b[0m processed records: 3440640\n",
      "[INFO]\u001b[0m processed records: 3448832\n",
      "[INFO]\u001b[0m processed records: 3457024\n",
      "[INFO]\u001b[0m processed records: 3465216\n",
      "[INFO]\u001b[0m processed records: 3473408\n",
      "[INFO]\u001b[0m processed records: 3481600\n",
      "[INFO]\u001b[0m processed records: 3489792\n",
      "[INFO]\u001b[0m processed records: 3497984\n",
      "[INFO]\u001b[0m processed records: 3506176\n",
      "[INFO]\u001b[0m processed records: 3514368\n",
      "[INFO]\u001b[0m processed records: 3522560\n",
      "[INFO]\u001b[0m processed records: 3530752\n",
      "[INFO]\u001b[0m processed records: 3538944\n",
      "[INFO]\u001b[0m processed records: 3547136\n",
      "[INFO]\u001b[0m processed records: 3555328\n",
      "[INFO]\u001b[0m processed records: 3563520\n",
      "[INFO]\u001b[0m processed records: 3571712\n",
      "[INFO]\u001b[0m processed records: 3579904\n",
      "[INFO]\u001b[0m processed records: 3588096\n",
      "[INFO]\u001b[0m processed records: 3596288\n",
      "[INFO]\u001b[0m processed records: 3604480\n",
      "[INFO]\u001b[0m processed records: 3612672\n",
      "[INFO]\u001b[0m processed records: 3620864\n",
      "[INFO]\u001b[0m processed records: 3629056\n",
      "[INFO]\u001b[0m processed records: 3637248\n",
      "[INFO]\u001b[0m processed records: 3645440\n",
      "[INFO]\u001b[0m processed records: 3653632\n",
      "[INFO]\u001b[0m processed records: 3661824\n",
      "[INFO]\u001b[0m processed records: 3670016\n",
      "[INFO]\u001b[0m processed records: 3678208\n",
      "[INFO]\u001b[0m processed records: 3686400\n",
      "[INFO]\u001b[0m processed records: 3694592\n",
      "[INFO]\u001b[0m processed records: 3702784\n",
      "[INFO]\u001b[0m processed records: 3710976\n",
      "[INFO]\u001b[0m processed records: 3719168\n",
      "[INFO]\u001b[0m processed records: 3727360\n",
      "[INFO]\u001b[0m processed records: 3735552\n",
      "[INFO]\u001b[0m processed records: 3743744\n",
      "[INFO]\u001b[0m processed records: 3751936\n",
      "[INFO]\u001b[0m processed records: 3760128\n",
      "[INFO]\u001b[0m processed records: 3768320\n",
      "[INFO]\u001b[0m processed records: 3776512\n",
      "[INFO]\u001b[0m processed records: 3784704\n",
      "[INFO]\u001b[0m processed records: 3792896\n",
      "[INFO]\u001b[0m processed records: 3801088\n",
      "[INFO]\u001b[0m processed records: 3809280\n",
      "[INFO]\u001b[0m processed records: 3817472\n",
      "[INFO]\u001b[0m processed records: 3825664\n",
      "[INFO]\u001b[0m processed records: 3833856\n",
      "[INFO]\u001b[0m processed records: 3842048\n",
      "[INFO]\u001b[0m processed records: 3850240\n",
      "[INFO]\u001b[0m processed records: 3858432\n",
      "[INFO]\u001b[0m processed records: 3866624\n",
      "[INFO]\u001b[0m processed records: 3874816\n",
      "[INFO]\u001b[0m processed records: 3883008\n",
      "[INFO]\u001b[0m processed records: 3891200\n",
      "[INFO]\u001b[0m processed records: 3899392\n",
      "[INFO]\u001b[0m processed records: 3907584\n",
      "[INFO]\u001b[0m processed records: 3915776\n",
      "[INFO]\u001b[0m processed records: 3923968\n",
      "[INFO]\u001b[0m processed records: 3932160\n",
      "[INFO]\u001b[0m processed records: 3940352\n",
      "[INFO]\u001b[0m processed records: 3948544\n",
      "[INFO]\u001b[0m processed records: 3956736\n",
      "[INFO]\u001b[0m processed records: 3964928\n",
      "[INFO]\u001b[0m processed records: 3973120\n",
      "[INFO]\u001b[0m processed records: 3981312\n",
      "[INFO]\u001b[0m processed records: 3989504\n",
      "[INFO]\u001b[0m processed records: 3997696\n",
      "[INFO]\u001b[0m processed records: 4005888\n",
      "[INFO]\u001b[0m processed records: 4014080\n",
      "[INFO]\u001b[0m processed records: 4022272\n",
      "[INFO]\u001b[0m processed records: 4030464\n",
      "[INFO]\u001b[0m processed records: 4038656\n",
      "[INFO]\u001b[0m processed records: 4046848\n",
      "[INFO]\u001b[0m processed records: 4055040\n",
      "[INFO]\u001b[0m processed records: 4063232\n",
      "[INFO]\u001b[0m processed records: 4071424\n",
      "[INFO]\u001b[0m processed records: 4079616\n",
      "[INFO]\u001b[0m processed records: 4087808\n",
      "[INFO]\u001b[0m processed records: 4096000\n",
      "[INFO]\u001b[0m processed records: 4104192\n",
      "[INFO]\u001b[0m processed records: 4112384\n",
      "[INFO]\u001b[0m processed records: 4120576\n",
      "[INFO]\u001b[0m processed records: 4128768\n",
      "[INFO]\u001b[0m processed records: 4136960\n",
      "[INFO]\u001b[0m processed records: 4145152\n",
      "[INFO]\u001b[0m processed records: 4153344\n",
      "[INFO]\u001b[0m processed records: 4161536\n",
      "[INFO]\u001b[0m processed records: 4169728\n",
      "[INFO]\u001b[0m processed records: 4177920\n",
      "[INFO]\u001b[0m processed records: 4186112\n",
      "[INFO]\u001b[0m processed records: 4194304\n",
      "[INFO]\u001b[0m processed records: 4202496\n",
      "[INFO]\u001b[0m processed records: 4210688\n",
      "[INFO]\u001b[0m processed records: 4218880\n",
      "[INFO]\u001b[0m processed records: 4227072\n",
      "[INFO]\u001b[0m processed records: 4235264\n",
      "[INFO]\u001b[0m processed records: 4243456\n",
      "[INFO]\u001b[0m processed records: 4251648\n",
      "[INFO]\u001b[0m processed records: 4259840\n",
      "[INFO]\u001b[0m processed records: 4268032\n",
      "[INFO]\u001b[0m processed records: 4276224\n",
      "[INFO]\u001b[0m processed records: 4284416\n",
      "[INFO]\u001b[0m processed records: 4292608\n",
      "[INFO]\u001b[0m processed records: 4300800\n",
      "[INFO]\u001b[0m processed records: 4308992\n",
      "[INFO]\u001b[0m processed records: 4317184\n",
      "[INFO]\u001b[0m processed records: 4325376\n",
      "[INFO]\u001b[0m processed records: 4333568\n",
      "[INFO]\u001b[0m processed records: 4341760\n",
      "[INFO]\u001b[0m processed records: 4349952\n",
      "[INFO]\u001b[0m processed records: 4358144\n",
      "[INFO]\u001b[0m processed records: 4366336\n",
      "[INFO]\u001b[0m processed records: 4374528\n",
      "[INFO]\u001b[0m processed records: 4382720\n",
      "[INFO]\u001b[0m processed records: 4390912\n",
      "[INFO]\u001b[0m processed records: 4399104\n",
      "[INFO]\u001b[0m processed records: 4407296\n",
      "[INFO]\u001b[0m processed records: 4415488\n",
      "[INFO]\u001b[0m processed records: 4423680\n",
      "[INFO]\u001b[0m processed records: 4431872\n",
      "[INFO]\u001b[0m processed records: 4440064\n",
      "[INFO]\u001b[0m processed records: 4448256\n",
      "[INFO]\u001b[0m processed records: 4456448\n",
      "[INFO]\u001b[0m processed records: 4464640\n",
      "[INFO]\u001b[0m processed records: 4472832\n",
      "[INFO]\u001b[0m processed records: 4481024\n",
      "[INFO]\u001b[0m processed records: 4489216\n",
      "[INFO]\u001b[0m processed records: 4497408\n",
      "[INFO]\u001b[0m processed records: 4505600\n",
      "[INFO]\u001b[0m processed records: 4513792\n",
      "[INFO]\u001b[0m processed records: 4521984\n",
      "[INFO]\u001b[0m processed records: 4530176\n",
      "[INFO]\u001b[0m processed records: 4538368\n",
      "[INFO]\u001b[0m processed records: 4546560\n",
      "[INFO]\u001b[0m processed records: 4554752\n",
      "[INFO]\u001b[0m processed records: 4562944\n",
      "[INFO]\u001b[0m processed records: 4571136\n",
      "[INFO]\u001b[0m processed records: 4579328\n",
      "[INFO]\u001b[0m processed records: 4587520\n",
      "[INFO]\u001b[0m processed records: 4595712\n",
      "[INFO]\u001b[0m processed records: 4603904\n",
      "[INFO]\u001b[0m processed records: 4612096\n",
      "[INFO]\u001b[0m processed records: 4620288\n",
      "[INFO]\u001b[0m processed records: 4628480\n",
      "[INFO]\u001b[0m processed records: 4636672\n",
      "[INFO]\u001b[0m processed records: 4644864\n",
      "[INFO]\u001b[0m processed records: 4653056\n",
      "[INFO]\u001b[0m processed records: 4661248\n",
      "[INFO]\u001b[0m processed records: 4669440\n",
      "[INFO]\u001b[0m processed records: 4677632\n",
      "[INFO]\u001b[0m processed records: 4685824\n",
      "[INFO]\u001b[0m processed records: 4694016\n",
      "[INFO]\u001b[0m processed records: 4702208\n",
      "[INFO]\u001b[0m processed records: 4710400\n",
      "[INFO]\u001b[0m processed records: 4718592\n",
      "[INFO]\u001b[0m processed records: 4726784\n",
      "[INFO]\u001b[0m processed records: 4734976\n",
      "[INFO]\u001b[0m processed records: 4743168\n",
      "[INFO]\u001b[0m processed records: 4751360\n",
      "[INFO]\u001b[0m processed records: 4759552\n",
      "[INFO]\u001b[0m processed records: 4767744\n",
      "[INFO]\u001b[0m processed records: 4775936\n",
      "[INFO]\u001b[0m processed records: 4784128\n",
      "[INFO]\u001b[0m processed records: 4792320\n",
      "[INFO]\u001b[0m processed records: 4800512\n",
      "[INFO]\u001b[0m processed records: 4808704\n",
      "[INFO]\u001b[0m processed records: 4816896\n",
      "[INFO]\u001b[0m processed records: 4825088\n",
      "[INFO]\u001b[0m processed records: 4833280\n",
      "[INFO]\u001b[0m processed records: 4841472\n",
      "[INFO]\u001b[0m processed records: 4849664\n",
      "[INFO]\u001b[0m processed records: 4857856\n",
      "[INFO]\u001b[0m processed records: 4866048\n",
      "[INFO]\u001b[0m processed records: 4874240\n",
      "[INFO]\u001b[0m processed records: 4882432\n",
      "[INFO]\u001b[0m processed records: 4890624\n",
      "[INFO]\u001b[0m processed records: 4898816\n",
      "[INFO]\u001b[0m processed records: 4907008\n",
      "[INFO]\u001b[0m processed records: 4915200\n",
      "[INFO]\u001b[0m processed records: 4923392\n",
      "[INFO]\u001b[0m processed records: 4931584\n",
      "[INFO]\u001b[0m processed records: 4939776\n",
      "[INFO]\u001b[0m processed records: 4947968\n",
      "[INFO]\u001b[0m processed records: 4956160\n",
      "[INFO]\u001b[0m processed records: 4964352\n",
      "[INFO]\u001b[0m processed records: 4972544\n",
      "[INFO]\u001b[0m processed records: 4980736\n",
      "[INFO]\u001b[0m processed records: 4988928\n",
      "[INFO]\u001b[0m processed records: 4997120\n",
      "[INFO]\u001b[0m processed records: 5005312\n",
      "[INFO]\u001b[0m processed records: 5013504\n",
      "[INFO]\u001b[0m processed records: 5021696\n",
      "[INFO]\u001b[0m processed records: 5029888\n",
      "[INFO]\u001b[0m processed records: 5038080\n",
      "[INFO]\u001b[0m processed records: 5046272\n",
      "[INFO]\u001b[0m processed records: 5054464\n",
      "[INFO]\u001b[0m processed records: 5062656\n",
      "[INFO]\u001b[0m processed records: 5070848\n",
      "[INFO]\u001b[0m processed records: 5079040\n",
      "[INFO]\u001b[0m processed records: 5087232\n",
      "[INFO]\u001b[0m processed records: 5095424\n",
      "[INFO]\u001b[0m processed records: 5103616\n",
      "[INFO]\u001b[0m processed records: 5111808\n",
      "[INFO]\u001b[0m processed records: 5120000\n",
      "[INFO]\u001b[0m processed records: 5128192\n",
      "[INFO]\u001b[0m processed records: 5136384\n",
      "[INFO]\u001b[0m processed records: 5144576\n",
      "[INFO]\u001b[0m processed records: 5152768\n",
      "[INFO]\u001b[0m processed records: 5160960\n",
      "[INFO]\u001b[0m processed records: 5169152\n",
      "[INFO]\u001b[0m processed records: 5177344\n",
      "[INFO]\u001b[0m processed records: 5185536\n",
      "[INFO]\u001b[0m processed records: 5193728\n",
      "[INFO]\u001b[0m processed records: 5201920\n",
      "[INFO]\u001b[0m processed records: 5210112\n",
      "[INFO]\u001b[0m processed records: 5218304\n",
      "[INFO]\u001b[0m processed records: 5226496\n",
      "[INFO]\u001b[0m processed records: 5234688\n",
      "[INFO]\u001b[0m processed records: 5242880\n",
      "[INFO]\u001b[0m processed records: 5251072\n",
      "[INFO]\u001b[0m processed records: 5259264\n",
      "[INFO]\u001b[0m processed records: 5267456\n",
      "[INFO]\u001b[0m processed records: 5275648\n",
      "[INFO]\u001b[0m processed records: 5283840\n",
      "[INFO]\u001b[0m processed records: 5292032\n",
      "[INFO]\u001b[0m processed records: 5300224\n",
      "[INFO]\u001b[0m processed records: 5308416\n",
      "[INFO]\u001b[0m processed records: 5316608\n",
      "[INFO]\u001b[0m processed records: 5324800\n",
      "[INFO]\u001b[0m processed records: 5332992\n",
      "[INFO]\u001b[0m processed records: 5341184\n",
      "[INFO]\u001b[0m processed records: 5349376\n",
      "[INFO]\u001b[0m processed records: 5357568\n",
      "[INFO]\u001b[0m processed records: 5365760\n",
      "[INFO]\u001b[0m processed records: 5373952\n",
      "[INFO]\u001b[0m processed records: 5382144\n",
      "[INFO]\u001b[0m processed records: 5390336\n",
      "[INFO]\u001b[0m processed records: 5398528\n",
      "[INFO]\u001b[0m processed records: 5406720\n",
      "[INFO]\u001b[0m processed records: 5414912\n",
      "[INFO]\u001b[0m processed records: 5423104\n",
      "[INFO]\u001b[0m processed records: 5431296\n",
      "[INFO]\u001b[0m processed records: 5439488\n",
      "[INFO]\u001b[0m processed records: 5447680\n",
      "[INFO]\u001b[0m processed records: 5455872\n",
      "[INFO]\u001b[0m processed records: 5464064\n",
      "[INFO]\u001b[0m processed records: 5472256\n",
      "[INFO]\u001b[0m processed records: 5480448\n",
      "[INFO]\u001b[0m processed records: 5488640\n",
      "[INFO]\u001b[0m processed records: 5496832\n",
      "[INFO]\u001b[0m processed records: 5505024\n",
      "[INFO]\u001b[0m processed records: 5513216\n",
      "[INFO]\u001b[0m processed records: 5521408\n",
      "[INFO]\u001b[0m processed records: 5529600\n",
      "[INFO]\u001b[0m processed records: 5537792\n",
      "[INFO]\u001b[0m processed records: 5545984\n",
      "[INFO]\u001b[0m processed records: 5554176\n",
      "[INFO]\u001b[0m processed records: 5562368\n",
      "[INFO]\u001b[0m processed records: 5570560\n",
      "[INFO]\u001b[0m processed records: 5578752\n",
      "[INFO]\u001b[0m processed records: 5586944\n",
      "[INFO]\u001b[0m processed records: 5595136\n",
      "[INFO]\u001b[0m processed records: 5603328\n",
      "[INFO]\u001b[0m processed records: 5611520\n",
      "[INFO]\u001b[0m processed records: 5619712\n",
      "[INFO]\u001b[0m processed records: 5627904\n",
      "[INFO]\u001b[0m processed records: 5636096\n",
      "[INFO]\u001b[0m processed records: 5644288\n",
      "[INFO]\u001b[0m processed records: 5652480\n",
      "[INFO]\u001b[0m processed records: 5660672\n",
      "[INFO]\u001b[0m processed records: 5668864\n",
      "[INFO]\u001b[0m processed records: 5677056\n",
      "[INFO]\u001b[0m processed records: 5685248\n",
      "[INFO]\u001b[0m processed records: 5693440\n",
      "[INFO]\u001b[0m processed records: 5701632\n",
      "[INFO]\u001b[0m processed records: 5709824\n",
      "[INFO]\u001b[0m processed records: 5718016\n",
      "[INFO]\u001b[0m processed records: 5726208\n",
      "[INFO]\u001b[0m processed records: 5734400\n",
      "[INFO]\u001b[0m processed records: 5742592\n",
      "[INFO]\u001b[0m processed records: 5750784\n",
      "[INFO]\u001b[0m processed records: 5758976\n",
      "[INFO]\u001b[0m processed records: 5767168\n",
      "[INFO]\u001b[0m processed records: 5775360\n",
      "[INFO]\u001b[0m processed records: 5783552\n",
      "[INFO]\u001b[0m processed records: 5791744\n",
      "[INFO]\u001b[0m processed records: 5799936\n",
      "[INFO]\u001b[0m processed records: 5808128\n",
      "[INFO]\u001b[0m processed records: 5816320\n",
      "[INFO]\u001b[0m processed records: 5824512\n",
      "[INFO]\u001b[0m processed records: 5832704\n",
      "[INFO]\u001b[0m processed records: 5840896\n",
      "[INFO]\u001b[0m processed records: 5849088\n",
      "[INFO]\u001b[0m processed records: 5857280\n",
      "[INFO]\u001b[0m processed records: 5865472\n",
      "[INFO]\u001b[0m processed records: 5873664\n",
      "[INFO]\u001b[0m processed records: 5881856\n",
      "[INFO]\u001b[0m processed records: 5890048\n",
      "[INFO]\u001b[0m processed records: 5898240\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Filter the NR virus taxid table\n",
    "seqkit fx2tab -n -i nr.virus.faa.gz > nr.virus.list.txt\n",
    "csvtk grep -t -H -f 1 -P nr.virus.list.txt virus.accession2taxid.ictv > nr.virus.accession2taxid.ictv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using the filtered FASTA, the ICTV taxdump, and the `virus.accession2taxid.ictv` tabular file, we will create a MMSeqs2 protein database with taxonomy information:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "createdb --dbtype 1 nr.virus.faa.gz virus_tax_db/virus_tax_db \n",
      "\n",
      "MMseqs Version:       \t15.6f452\n",
      "Database type         \t1\n",
      "Shuffle input database\ttrue\n",
      "Createdb mode         \t0\n",
      "Write lookup file     \t1\n",
      "Offset of numeric ids \t0\n",
      "Compressed            \t0\n",
      "Verbosity             \t3\n",
      "\n",
      "Converting sequences\n",
      "[===================================================================================================\t1 Mio. sequences processed\n",
      "===================================================================================================\t2 Mio. sequences processed\n",
      "===================================================================================================\t3 Mio. sequences processed\n",
      "===================================================================================================\t4 Mio. sequences processed\n",
      "===================================================================================================\t5 Mio. sequences processed\n",
      "==========================================================================================\n",
      "Time for merging to virus_tax_db_h: 0h 0m 1s 292ms\n",
      "Time for merging to virus_tax_db: 0h 0m 5s 937ms\n",
      "Database type: Aminoacid\n",
      "Time for processing: 0h 0m 29s 511ms\n",
      "Create directory tmp\n",
      "createtaxdb virus_tax_db/virus_tax_db tmp --ncbi-tax-dump ictv-taxdump --tax-mapping-file nr.virus.accession2taxid.ictv \n",
      "\n",
      "MMseqs Version:        \t15.6f452\n",
      "NCBI tax dump directory\tictv-taxdump\n",
      "Taxonomy mapping file  \tnr.virus.accession2taxid.ictv\n",
      "Taxonomy mapping mode  \t0\n",
      "Taxonomy db mode       \t1\n",
      "Threads                \t32\n",
      "Verbosity              \t3\n",
      "\n",
      "Loading nodes file ... Done, got 18683 nodes\n",
      "Loading merged file ... Done, added 0 merged nodes.\n",
      "Loading names file ... Done\n",
      "Init RMQ ...Done\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Create the MMSeqs2 database\n",
    "mkdir virus_tax_db\n",
    "mmseqs createdb --dbtype 1 nr.virus.faa.gz virus_tax_db/virus_tax_db\n",
    "mmseqs createtaxdb virus_tax_db/virus_tax_db tmp --ncbi-tax-dump ictv-taxdump --tax-mapping-file nr.virus.accession2taxid.ictv\n",
    "rm -rf tmp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, to assign taxonomy to sequences in the ICTV \"\"\"chellenge\"\"\" (quatation as I hope it would make the powers to be realize half of the \"higher level taxa\" are related by hand waving, which is fun we all like to be a little philosophical, I guess...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# wget https://github.com/ICTV-VBEG/ICTV-TaxonomyChallenge/raw/main/dataset/dataset_challenge.tar.gz?download= -O dataset_challenge.tar.gz -o /dev/null\n",
    "tar -xzf dataset_challenge.tar.gz # considering all of these are non segmented, i.e. 1 contig per input fasta file - this is a really annoying way to bundle them up (why not 1 multifasta file?)\n",
    "seqkit stats dataset_challenge/*fasta > stats.info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we'll concatenate the many single fasta files, and use mmseqs taxonomy on that."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "taxonomy ictv_chellenge/mmdb virus_tax_db/virus_tax_db taxonomy_results/resdb2 tmp -e 1e-5 -s 6 --blacklist  --tax-lineage 1 --threads 14 --lca-ranks superkingdom,kingdom,subkingdom,superphylum,phylum,subphylum,class,subclass,order,suborder,family,subfamily,genus,subgenus,species \n",
      "\n",
      "MMseqs Version:                        \t15.6f452\n",
      "ORF filter                             \t1\n",
      "ORF filter e-value                     \t100\n",
      "ORF filter sensitivity                 \t2\n",
      "LCA mode                               \t3\n",
      "Taxonomy output mode                   \t0\n",
      "Majority threshold                     \t0.5\n",
      "Vote mode                              \t1\n",
      "LCA ranks                              \tsuperkingdom,kingdom,subkingdom,superphylum,phylum,subphylum,class,subclass,order,suborder,family,subfamily,genus,subgenus,species\n",
      "Column with taxonomic lineage          \t1\n",
      "Compressed                             \t0\n",
      "Threads                                \t14\n",
      "Verbosity                              \t3\n",
      "Taxon blacklist                        \t\n",
      "Substitution matrix                    \taa:blosum62.out,nucl:nucleotide.out\n",
      "Add backtrace                          \tfalse\n",
      "Alignment mode                         \t1\n",
      "Alignment mode                         \t0\n",
      "Allow wrapped scoring                  \tfalse\n",
      "E-value threshold                      \t1e-05\n",
      "Seq. id. threshold                     \t0\n",
      "Min alignment length                   \t0\n",
      "Seq. id. mode                          \t0\n",
      "Alternative alignments                 \t0\n",
      "Coverage threshold                     \t0\n",
      "Coverage mode                          \t0\n",
      "Max sequence length                    \t65535\n",
      "Compositional bias                     \t1\n",
      "Compositional bias                     \t1\n",
      "Max reject                             \t5\n",
      "Max accept                             \t30\n",
      "Include identical seq. id.             \tfalse\n",
      "Preload mode                           \t0\n",
      "Pseudo count a                         \tsubstitution:1.100,context:1.400\n",
      "Pseudo count b                         \tsubstitution:4.100,context:5.800\n",
      "Score bias                             \t0\n",
      "Realign hits                           \tfalse\n",
      "Realign score bias                     \t-0.2\n",
      "Realign max seqs                       \t2147483647\n",
      "Correlation score weight               \t0\n",
      "Gap open cost                          \taa:11,nucl:5\n",
      "Gap extension cost                     \taa:1,nucl:2\n",
      "Zdrop                                  \t40\n",
      "Seed substitution matrix               \taa:VTML80.out,nucl:nucleotide.out\n",
      "Sensitivity                            \t6\n",
      "k-mer length                           \t0\n",
      "Target search mode                     \t0\n",
      "k-score                                \tseq:2147483647,prof:2147483647\n",
      "Alphabet size                          \taa:21,nucl:5\n",
      "Max results per query                  \t300\n",
      "Split database                         \t0\n",
      "Split mode                             \t2\n",
      "Split memory limit                     \t0\n",
      "Diagonal scoring                       \ttrue\n",
      "Exact k-mer matching                   \t0\n",
      "Mask residues                          \t1\n",
      "Mask residues probability              \t0.9\n",
      "Mask lower case residues               \t0\n",
      "Minimum diagonal score                 \t15\n",
      "Selected taxa                          \t\n",
      "Spaced k-mers                          \t1\n",
      "Spaced k-mer pattern                   \t\n",
      "Local temporary path                   \t\n",
      "Rescore mode                           \t0\n",
      "Remove hits by seq. id. and coverage   \tfalse\n",
      "Sort results                           \t0\n",
      "Mask profile                           \t1\n",
      "Profile E-value threshold              \t0.001\n",
      "Global sequence weighting              \tfalse\n",
      "Allow deletions                        \tfalse\n",
      "Filter MSA                             \t1\n",
      "Use filter only at N seqs              \t0\n",
      "Maximum seq. id. threshold             \t0.9\n",
      "Minimum seq. id.                       \t0.0\n",
      "Minimum score per column               \t-20\n",
      "Minimum coverage                       \t0\n",
      "Select N most diverse seqs             \t1000\n",
      "Pseudo count mode                      \t0\n",
      "Min codons in orf                      \t30\n",
      "Max codons in length                   \t32734\n",
      "Max orf gaps                           \t2147483647\n",
      "Contig start mode                      \t2\n",
      "Contig end mode                        \t2\n",
      "Orf start mode                         \t1\n",
      "Forward frames                         \t1,2,3\n",
      "Reverse frames                         \t1,2,3\n",
      "Translation table                      \t1\n",
      "Translate orf                          \t0\n",
      "Use all table starts                   \tfalse\n",
      "Offset of numeric ids                  \t0\n",
      "Create lookup                          \t0\n",
      "Add orf stop                           \tfalse\n",
      "Overlap between sequences              \t0\n",
      "Sequence split mode                    \t1\n",
      "Header split mode                      \t0\n",
      "Chain overlapping alignments           \t0\n",
      "Merge query                            \t1\n",
      "Search type                            \t0\n",
      "Prefilter mode                         \t0\n",
      "Exhaustive search mode                 \tfalse\n",
      "Filter results during exhaustive search\t0\n",
      "Strand selection                       \t1\n",
      "LCA search mode                        \tfalse\n",
      "Disk space limit                       \t0\n",
      "MPI runner                             \t\n",
      "Force restart with latest tmp          \tfalse\n",
      "Remove temporary files                 \tfalse\n",
      "\n",
      "taxonomy tmp/78675634633872952/orfs_filter virus_tax_db/virus_tax_db tmp/78675634633872952/orfs_tax tmp/78675634633872952/tmp_taxonomy --tax-output-mode 2 --lca-ranks superkingdom,kingdom,subkingdom,superphylum,phylum,subphylum,class,subclass,order,suborder,family,subfamily,genus,subgenus,species --tax-lineage 0 --threads 14 --alignment-mode 1 -e 1e-05 --max-rejected 5 --max-accept 30 -s 6 --spaced-kmer-mode 1 --min-length 30 --max-length 32734 --orf-start-mode 1 \n",
      "\n",
      "search tmp/78675634633872952/orfs_filter virus_tax_db/virus_tax_db tmp/78675634633872952/tmp_taxonomy/11645834057794403919/first tmp/78675634633872952/tmp_taxonomy/11645834057794403919/tmp_hsp1 --alignment-mode 1 -e 1e-05 --max-rejected 5 --max-accept 30 --threads 14 -s 6 --spaced-kmer-mode 1 --min-length 30 --max-length 32734 --orf-start-mode 1 --lca-search 1 \n",
      "\n",
      "prefilter tmp/78675634633872952/orfs_filter virus_tax_db/virus_tax_db tmp/78675634633872952/tmp_taxonomy/11645834057794403919/tmp_hsp1/3957465038754195884/pref_0 --sub-mat 'aa:blosum62.out,nucl:nucleotide.out' --seed-sub-mat 'aa:VTML80.out,nucl:nucleotide.out' -k 0 --target-search-mode 0 --k-score seq:2147483647,prof:2147483647 --alph-size aa:21,nucl:5 --max-seq-len 65535 --max-seqs 300 --split 0 --split-mode 2 --split-memory-limit 0 -c 0 --cov-mode 0 --comp-bias-corr 1 --comp-bias-corr-scale 1 --diag-score 1 --exact-kmer-matching 0 --mask 1 --mask-prob 0.9 --mask-lower-case 0 --min-ungapped-score 15 --add-self-matches 0 --spaced-kmer-mode 1 --db-load-mode 0 --pca substitution:1.100,context:1.400 --pcb substitution:4.100,context:5.800 --threads 14 --compressed 0 -v 3 -s 6.0 \n",
      "\n",
      "Query database size: 1292606 type: Aminoacid\n",
      "Estimated memory consumption: 16G\n",
      "Target database size: 5900899 type: Aminoacid\n",
      "Index table k-mer threshold: 109 at k-mer size 6 \n",
      "Index table: counting k-mers\n",
      "[=================================================================] 5.90M 18s 812ms\n",
      "Index table: Masked residues: 15182170\n",
      "Index table: fill\n",
      "[=================================================================] 5.90M 29s 810ms\n",
      "Index statistics\n",
      "Entries:          1644466474\n",
      "DB size:          9897 MB\n",
      "Avg k-mer size:   25.694789\n",
      "Top 10 k-mers\n",
      "    PKLDDY\t179243\n",
      "    ESAIRL\t166363\n",
      "    RFEGRR\t166320\n",
      "    LKREEG\t166220\n",
      "    IERRAW\t166201\n",
      "    RAITFT\t166158\n",
      "    DESAKT\t166151\n",
      "    TRFIEM\t165332\n",
      "    ETKACT\t165245\n",
      "    FSTEAT\t165187\n",
      "Time for index table init: 0h 0m 53s 38ms\n",
      "Process prefiltering step 1 of 1\n",
      "\n",
      "k-mer similarity threshold: 109\n",
      "Starting prefiltering scores calculation (step 1 of 1)\n",
      "Query db start 1 to 1292606\n",
      "Target db start 1 to 5900899\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cat dataset_challenge/*fasta > ictv_chellenge.fasta\n",
    "mkdir ictv_chellenge taxonomy_results/\n",
    "mmseqs createdb ictv_chellenge.fasta ictv_chellenge/mmdb\n",
    "\n",
    "mmseqs taxonomy ictv_chellenge/mmdb virus_tax_db/virus_tax_db taxonomy_results/resdb2 tmp -e 0.001 -s 7 --blacklist \"\" --tax-lineage 1 --threads 14 --lca-ranks superkingdom,kingdom,subkingdom,superphylum,phylum,subphylum,class,subclass,order,suborder,family,subfamily,genus,subgenus,species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mmseqs createtsv ictv_chellenge/mmdb taxonomy_results/resdb2 taxonomyResult.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we'll read the results and format them as the ICTV chellenge wants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "import polars as pl\n",
    "taxonomyResult = pl.read_csv('taxonomyResult.tsv', separator='\\t', has_header=False,new_columns=\"qseqid,LCA_ncbi_taxid,rank,taxon_name,protein_fragments,protein_fragments_retained,protein_fragments_assigned,protein_fragments_in_agreement_with_lca,lineage\".split(',')) \n",
    "# Rename columns for clarity\n",
    "taxonomyResult_mini = taxonomyResult.rename(\n",
    "    {'qseqid':\"SequenceID\",'LCA_ncbi_taxid' : 'tax_id',\"taxon_name\" : \"taxname\",\"protein_fragments_in_agreement_with_lca\" : \"score\"}).select(pl.col([\"SequenceID\",\"rank\",\"tax_id\",\"taxname\",\"score\",\"lineage\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [],
   "source": [
    "def drop_null_columns(df):\n",
    "    return df.drop([col for col in df.columns if df[col].is_null().all()])\n",
    "\n",
    "def capitalize_column_names(df: pl.DataFrame) -> pl.DataFrame:\n",
    "    # Create a dictionary mapping old column names to capitalized names\n",
    "    new_names = {col: col.capitalize() for col in df.columns}\n",
    "    \n",
    "    # Rename the columns using the dictionary\n",
    "    return df.rename(new_names)\n",
    "\n",
    "def order_columns_alphabetically(df: pl.DataFrame) -> pl.DataFrame:\n",
    "    # Get the list of column names and sort them alphabetically\n",
    "    sorted_columns = sorted(df.columns)\n",
    "    \n",
    "    # Use select to reorder the columns\n",
    "    return df.select(sorted_columns)\n",
    "\n",
    "taxonomyResult_mini= capitalize_column_names(taxonomyResult_mini)\n",
    "taxonomyResult_mini= order_columns_alphabetically(taxonomyResult_mini)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (52_193, 6)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Lineage</th><th>Rank</th><th>Score</th><th>Sequenceid</th><th>Tax_id</th><th>Taxname</th></tr><tr><td>str</td><td>str</td><td>f64</td><td>str</td><td>i64</td><td>str</td></tr></thead><tbody><tr><td>&quot;-_Duplodnaviria;k_Heunggongvir…</td><td>&quot;genus&quot;</td><td>0.59</td><td>&quot;ICTVTaxoChallenge_45380&quot;</td><td>7049</td><td>&quot;Camvirus&quot;</td></tr><tr><td>&quot;-_Riboviria;k_Orthornavirae;p_…</td><td>&quot;species&quot;</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_454570&quot;</td><td>2257</td><td>&quot;Emaravirus kiwii&quot;</td></tr><tr><td>&quot;-_Riboviria;k_Orthornavirae;p_…</td><td>&quot;family&quot;</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_454980&quot;</td><td>243</td><td>&quot;Steitzviridae&quot;</td></tr><tr><td>&quot;-_Monodnaviria;k_Shotokuvirae;…</td><td>&quot;genus&quot;</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_455533&quot;</td><td>1009</td><td>&quot;Tetraparvovirus&quot;</td></tr><tr><td>&quot;-_Riboviria;k_Orthornavirae;p_…</td><td>&quot;genus&quot;</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_455965&quot;</td><td>8157</td><td>&quot;Phlebovirus&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;-_Duplodnaviria;k_Heunggongvir…</td><td>&quot;genus&quot;</td><td>0.7</td><td>&quot;ICTVTaxoChallenge_693184&quot;</td><td>11874</td><td>&quot;Slopekvirus&quot;</td></tr><tr><td>&quot;-_Monodnaviria;k_Shotokuvirae;…</td><td>&quot;species&quot;</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_693699&quot;</td><td>3634</td><td>&quot;Etapapillomavirus 1&quot;</td></tr><tr><td>&quot;-_Riboviria;k_Orthornavirae;p_…</td><td>&quot;genus&quot;</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_694175&quot;</td><td>9472</td><td>&quot;Rotavirus&quot;</td></tr><tr><td>&quot;-_Duplodnaviria;k_Heunggongvir…</td><td>&quot;species&quot;</td><td>0.51</td><td>&quot;ICTVTaxoChallenge_694718&quot;</td><td>5351</td><td>&quot;Timshelvirus droogsarmy&quot;</td></tr><tr><td>&quot;-_Riboviria;k_Pararnavirae;p_A…</td><td>&quot;species&quot;</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_695160&quot;</td><td>5601</td><td>&quot;Badnavirus deltavirgamusae&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (52_193, 6)\n",
       "┌───────────────────────┬─────────┬───────┬───────────────────────┬────────┬───────────────────────┐\n",
       "│ Lineage               ┆ Rank    ┆ Score ┆ Sequenceid            ┆ Tax_id ┆ Taxname               │\n",
       "│ ---                   ┆ ---     ┆ ---   ┆ ---                   ┆ ---    ┆ ---                   │\n",
       "│ str                   ┆ str     ┆ f64   ┆ str                   ┆ i64    ┆ str                   │\n",
       "╞═══════════════════════╪═════════╪═══════╪═══════════════════════╪════════╪═══════════════════════╡\n",
       "│ -_Duplodnaviria;k_Heu ┆ genus   ┆ 0.59  ┆ ICTVTaxoChallenge_453 ┆ 7049   ┆ Camvirus              │\n",
       "│ nggongvir…            ┆         ┆       ┆ 80                    ┆        ┆                       │\n",
       "│ -_Riboviria;k_Orthorn ┆ species ┆ 1.0   ┆ ICTVTaxoChallenge_454 ┆ 2257   ┆ Emaravirus kiwii      │\n",
       "│ avirae;p_…            ┆         ┆       ┆ 570                   ┆        ┆                       │\n",
       "│ -_Riboviria;k_Orthorn ┆ family  ┆ 1.0   ┆ ICTVTaxoChallenge_454 ┆ 243    ┆ Steitzviridae         │\n",
       "│ avirae;p_…            ┆         ┆       ┆ 980                   ┆        ┆                       │\n",
       "│ -_Monodnaviria;k_Shot ┆ genus   ┆ 1.0   ┆ ICTVTaxoChallenge_455 ┆ 1009   ┆ Tetraparvovirus       │\n",
       "│ okuvirae;…            ┆         ┆       ┆ 533                   ┆        ┆                       │\n",
       "│ -_Riboviria;k_Orthorn ┆ genus   ┆ 1.0   ┆ ICTVTaxoChallenge_455 ┆ 8157   ┆ Phlebovirus           │\n",
       "│ avirae;p_…            ┆         ┆       ┆ 965                   ┆        ┆                       │\n",
       "│ …                     ┆ …       ┆ …     ┆ …                     ┆ …      ┆ …                     │\n",
       "│ -_Duplodnaviria;k_Heu ┆ genus   ┆ 0.7   ┆ ICTVTaxoChallenge_693 ┆ 11874  ┆ Slopekvirus           │\n",
       "│ nggongvir…            ┆         ┆       ┆ 184                   ┆        ┆                       │\n",
       "│ -_Monodnaviria;k_Shot ┆ species ┆ 1.0   ┆ ICTVTaxoChallenge_693 ┆ 3634   ┆ Etapapillomavirus 1   │\n",
       "│ okuvirae;…            ┆         ┆       ┆ 699                   ┆        ┆                       │\n",
       "│ -_Riboviria;k_Orthorn ┆ genus   ┆ 1.0   ┆ ICTVTaxoChallenge_694 ┆ 9472   ┆ Rotavirus             │\n",
       "│ avirae;p_…            ┆         ┆       ┆ 175                   ┆        ┆                       │\n",
       "│ -_Duplodnaviria;k_Heu ┆ species ┆ 0.51  ┆ ICTVTaxoChallenge_694 ┆ 5351   ┆ Timshelvirus          │\n",
       "│ nggongvir…            ┆         ┆       ┆ 718                   ┆        ┆ droogsarmy            │\n",
       "│ -_Riboviria;k_Pararna ┆ species ┆ 1.0   ┆ ICTVTaxoChallenge_695 ┆ 5601   ┆ Badnavirus            │\n",
       "│ virae;p_A…            ┆         ┆       ┆ 160                   ┆        ┆ deltavirgamusae       │\n",
       "└───────────────────────┴─────────┴───────┴───────────────────────┴────────┴───────────────────────┘"
      ]
     },
     "execution_count": 198,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ictv_vmr = pl.read_csv(\"ictv.csv\")\n",
    "ictv_vmr_mini = ictv_vmr.select(pl.col(\"Realm,Kingdom,Subkingdom,Phylum,Subphylum,Class,Subclass,Order,Suborder,Family,Subfamily,Genus,Subgenus,Species\".split(','))).unique()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "pivoted = taxonomyResult_mini.pivot(\n",
    "    values=[ 'Taxname'],\n",
    "    index='Sequenceid',\n",
    "    on='Rank'\n",
    ")\n",
    "# pivoted.rename({\"no_rank\":\"Realm\"})\n",
    "pivoted[\"no rank\"].unique()\n",
    "\n",
    "pivoted = order_columns_alphabetically(capitalize_column_names(pivoted))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "metadata": {},
   "outputs": [],
   "source": [
    "final=[]\n",
    "column_order=[\"SequenceID\", set(ictv_vmr_mini.columns).intersection(pivoted.columns)]\n",
    "for taxrank in ictv_vmr_mini.columns:\n",
    "    if taxrank in pivoted.columns:\n",
    "        pivoted_taxrank_only = pivoted.select([pl.col(taxrank),\"Sequenceid\"] )\n",
    "        pivoted_taxrank_only = pivoted_taxrank_only.drop([col for col in pivoted_taxrank_only.columns if col not in [taxrank,\"Sequenceid\",\"Score\"] ])\n",
    "        pivoted_taxrank_only = pivoted_taxrank_only.join(ictv_vmr_mini, how=\"left\", on=taxrank)\n",
    "        pivoted_taxrank_only = order_columns_alphabetically(pivoted_taxrank_only)\n",
    "        final.append(pivoted_taxrank_only)\n",
    "final=pl.concat(final, how=\"vertical\").unique()\n",
    "final=final.unique(subset=[\"Sequenceid\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [],
   "source": [
    "tiny = taxonomyResult_mini.select([\"Sequenceid\",\"Score\"])\n",
    "final = tiny.join(final, how=\"left\", on=\"Sequenceid\")\n",
    "# duplicate the score column for each rank automatically\n",
    "for taxrank in ictv_vmr_mini.columns:\n",
    "    if taxrank in final.columns:\n",
    "        final = final.with_columns(pl.col(\"Score\").alias(f\"Score_{taxrank}\"))\n",
    "\n",
    "# clean up the output, add fasta headers of contig names not in the mmseqs output\n",
    "final = final.rename({\"Sequenceid\":\"SequenceID\"}) \n",
    "final = final.drop([\"Score\"])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_fasta = \"ictv_chellenge.fasta\"\n",
    "contig_names = {}\n",
    "with open(input_fasta, 'r') as fasta_file:\n",
    "    for line in fasta_file:\n",
    "        if line.startswith('>'):\n",
    "            contig_name = line.strip()[1:]\n",
    "            contig_names[contig_name] = True\n",
    "# add the missing contig names to the final dataframe,\n",
    "missing_contigs = pl.DataFrame({\"SequenceID\": list(set(contig_names.keys()) - set(final[\"SequenceID\"]))})\n",
    "# set scores of all tanks to 1, and set all rank to \"unclassified\"\n",
    "for taxrank in ictv_vmr_mini.columns:\n",
    "    if taxrank in final.columns:\n",
    "        missing_contigs = missing_contigs.with_columns(pl.lit(\"unclassified\").alias(taxrank))\n",
    "        # set scores of all tanks to 1\n",
    "        missing_contigs = missing_contigs.with_columns(pl.lit(1.0).alias(f\"Score_{taxrank}\"))\n",
    "\n",
    "final2 = pl.concat([order_columns_alphabetically(final), order_columns_alphabetically(missing_contigs) ])\n",
    "\n",
    "\n",
    "final2.write_csv(\"submission_ctcv.csv\", separator=\",\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (59_907, 29)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>Class</th><th>Family</th><th>Genus</th><th>Kingdom</th><th>Order</th><th>Phylum</th><th>Realm</th><th>Score_Class</th><th>Score_Family</th><th>Score_Genus</th><th>Score_Kingdom</th><th>Score_Order</th><th>Score_Phylum</th><th>Score_Realm</th><th>Score_Species</th><th>Score_Subclass</th><th>Score_Subfamily</th><th>Score_Subgenus</th><th>Score_Subkingdom</th><th>Score_Suborder</th><th>Score_Subphylum</th><th>SequenceID</th><th>Species</th><th>Subclass</th><th>Subfamily</th><th>Subgenus</th><th>Subkingdom</th><th>Suborder</th><th>Subphylum</th></tr><tr><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>f64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;Caudoviricetes&quot;</td><td>null</td><td>&quot;Camvirus&quot;</td><td>&quot;Heunggongvirae&quot;</td><td>null</td><td>&quot;Uroviricota&quot;</td><td>&quot;Duplodnaviria&quot;</td><td>0.59</td><td>0.59</td><td>0.59</td><td>0.59</td><td>0.59</td><td>0.59</td><td>0.59</td><td>0.59</td><td>0.59</td><td>0.59</td><td>0.59</td><td>0.59</td><td>0.59</td><td>0.59</td><td>&quot;ICTVTaxoChallenge_45380&quot;</td><td>&quot;Camvirus CAM&quot;</td><td>null</td><td>&quot;Arquatrovirinae&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_454570&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>&quot;Leviviricetes&quot;</td><td>&quot;Steitzviridae&quot;</td><td>&quot;Bahrchivirus&quot;</td><td>&quot;Orthornavirae&quot;</td><td>&quot;Timlovirales&quot;</td><td>&quot;Lenarviricota&quot;</td><td>&quot;Riboviria&quot;</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_454980&quot;</td><td>&quot;Bahrchivirus agrivivens&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>&quot;Quintoviricetes&quot;</td><td>&quot;Parvoviridae&quot;</td><td>&quot;Tetraparvovirus&quot;</td><td>&quot;Shotokuvirae&quot;</td><td>&quot;Piccovirales&quot;</td><td>&quot;Cossaviricota&quot;</td><td>&quot;Monodnaviria&quot;</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_455533&quot;</td><td>&quot;Tetraparvovirus rodent1&quot;</td><td>null</td><td>&quot;Parvovirinae&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>&quot;Bunyaviricetes&quot;</td><td>&quot;Phenuiviridae&quot;</td><td>&quot;Phlebovirus&quot;</td><td>&quot;Orthornavirae&quot;</td><td>&quot;Hareavirales&quot;</td><td>&quot;Negarnaviricota&quot;</td><td>&quot;Riboviria&quot;</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_455965&quot;</td><td>&quot;Phlebovirus mukawaense&quot;</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>&quot;Polyploviricotina&quot;</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_966767&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td></tr><tr><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_60784&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td></tr><tr><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_504800&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td></tr><tr><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_425649&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td></tr><tr><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>1.0</td><td>&quot;ICTVTaxoChallenge_490100&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td><td>&quot;unclassified&quot;</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (59_907, 29)\n",
       "┌───────────┬───────────┬───────────┬───────────┬───┬───────────┬───────────┬───────────┬──────────┐\n",
       "│ Class     ┆ Family    ┆ Genus     ┆ Kingdom   ┆ … ┆ Subgenus  ┆ Subkingdo ┆ Suborder  ┆ Subphylu │\n",
       "│ ---       ┆ ---       ┆ ---       ┆ ---       ┆   ┆ ---       ┆ m         ┆ ---       ┆ m        │\n",
       "│ str       ┆ str       ┆ str       ┆ str       ┆   ┆ str       ┆ ---       ┆ str       ┆ ---      │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆ str       ┆           ┆ str      │\n",
       "╞═══════════╪═══════════╪═══════════╪═══════════╪═══╪═══════════╪═══════════╪═══════════╪══════════╡\n",
       "│ Caudoviri ┆ null      ┆ Camvirus  ┆ Heunggong ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│ cetes     ┆           ┆           ┆ virae     ┆   ┆           ┆           ┆           ┆          │\n",
       "│ null      ┆ null      ┆ null      ┆ null      ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│ Leviviric ┆ Steitzvir ┆ Bahrchivi ┆ Orthornav ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│ etes      ┆ idae      ┆ rus       ┆ irae      ┆   ┆           ┆           ┆           ┆          │\n",
       "│ Quintovir ┆ Parvoviri ┆ Tetraparv ┆ Shotokuvi ┆ … ┆ null      ┆ null      ┆ null      ┆ null     │\n",
       "│ icetes    ┆ dae       ┆ ovirus    ┆ rae       ┆   ┆           ┆           ┆           ┆          │\n",
       "│ Bunyaviri ┆ Phenuivir ┆ Phlebovir ┆ Orthornav ┆ … ┆ null      ┆ null      ┆ null      ┆ Polyplov │\n",
       "│ cetes     ┆ idae      ┆ us        ┆ irae      ┆   ┆           ┆           ┆           ┆ iricotin │\n",
       "│           ┆           ┆           ┆           ┆   ┆           ┆           ┆           ┆ a        │\n",
       "│ …         ┆ …         ┆ …         ┆ …         ┆ … ┆ …         ┆ …         ┆ …         ┆ …        │\n",
       "│ unclassif ┆ unclassif ┆ unclassif ┆ unclassif ┆ … ┆ unclassif ┆ unclassif ┆ unclassif ┆ unclassi │\n",
       "│ ied       ┆ ied       ┆ ied       ┆ ied       ┆   ┆ ied       ┆ ied       ┆ ied       ┆ fied     │\n",
       "│ unclassif ┆ unclassif ┆ unclassif ┆ unclassif ┆ … ┆ unclassif ┆ unclassif ┆ unclassif ┆ unclassi │\n",
       "│ ied       ┆ ied       ┆ ied       ┆ ied       ┆   ┆ ied       ┆ ied       ┆ ied       ┆ fied     │\n",
       "│ unclassif ┆ unclassif ┆ unclassif ┆ unclassif ┆ … ┆ unclassif ┆ unclassif ┆ unclassif ┆ unclassi │\n",
       "│ ied       ┆ ied       ┆ ied       ┆ ied       ┆   ┆ ied       ┆ ied       ┆ ied       ┆ fied     │\n",
       "│ unclassif ┆ unclassif ┆ unclassif ┆ unclassif ┆ … ┆ unclassif ┆ unclassif ┆ unclassif ┆ unclassi │\n",
       "│ ied       ┆ ied       ┆ ied       ┆ ied       ┆   ┆ ied       ┆ ied       ┆ ied       ┆ fied     │\n",
       "│ unclassif ┆ unclassif ┆ unclassif ┆ unclassif ┆ … ┆ unclassif ┆ unclassif ┆ unclassif ┆ unclassi │\n",
       "│ ied       ┆ ied       ┆ ied       ┆ ied       ┆   ┆ ied       ┆ ied       ┆ ied       ┆ fied     │\n",
       "└───────────┴───────────┴───────────┴───────────┴───┴───────────┴───────────┴───────────┴──────────┘"
      ]
     },
     "execution_count": 256,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div><style>\n",
       ".dataframe > thead > tr,\n",
       ".dataframe > tbody > tr {\n",
       "  text-align: right;\n",
       "  white-space: pre-wrap;\n",
       "}\n",
       "</style>\n",
       "<small>shape: (52_193, 14)</small><table border=\"1\" class=\"dataframe\"><thead><tr><th>SequenceID</th><th>rank</th><th>tax_id</th><th>taxname</th><th>score</th><th>lineage</th><th>realm</th><th>kingdom</th><th>phylum</th><th>class</th><th>order</th><th>family</th><th>genus</th><th>species</th></tr><tr><td>str</td><td>str</td><td>i64</td><td>str</td><td>f64</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td><td>str</td></tr></thead><tbody><tr><td>&quot;ICTVTaxoChallenge_45380&quot;</td><td>&quot;genus&quot;</td><td>7049</td><td>&quot;Camvirus&quot;</td><td>0.59</td><td>&quot;-_Duplodnaviria;k_Heunggongvir…</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>&quot;ICTVTaxoChallenge_454570&quot;</td><td>&quot;species&quot;</td><td>2257</td><td>&quot;Emaravirus kiwii&quot;</td><td>1.0</td><td>&quot;-_Riboviria;k_Orthornavirae;p_…</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>&quot;ICTVTaxoChallenge_454980&quot;</td><td>&quot;family&quot;</td><td>243</td><td>&quot;Steitzviridae&quot;</td><td>1.0</td><td>&quot;-_Riboviria;k_Orthornavirae;p_…</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>&quot;ICTVTaxoChallenge_455533&quot;</td><td>&quot;genus&quot;</td><td>1009</td><td>&quot;Tetraparvovirus&quot;</td><td>1.0</td><td>&quot;-_Monodnaviria;k_Shotokuvirae;…</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>&quot;ICTVTaxoChallenge_455965&quot;</td><td>&quot;genus&quot;</td><td>8157</td><td>&quot;Phlebovirus&quot;</td><td>1.0</td><td>&quot;-_Riboviria;k_Orthornavirae;p_…</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td><td>&hellip;</td></tr><tr><td>&quot;ICTVTaxoChallenge_693184&quot;</td><td>&quot;genus&quot;</td><td>11874</td><td>&quot;Slopekvirus&quot;</td><td>0.7</td><td>&quot;-_Duplodnaviria;k_Heunggongvir…</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>&quot;ICTVTaxoChallenge_693699&quot;</td><td>&quot;species&quot;</td><td>3634</td><td>&quot;Etapapillomavirus 1&quot;</td><td>1.0</td><td>&quot;-_Monodnaviria;k_Shotokuvirae;…</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>&quot;ICTVTaxoChallenge_694175&quot;</td><td>&quot;genus&quot;</td><td>9472</td><td>&quot;Rotavirus&quot;</td><td>1.0</td><td>&quot;-_Riboviria;k_Orthornavirae;p_…</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>&quot;ICTVTaxoChallenge_694718&quot;</td><td>&quot;species&quot;</td><td>5351</td><td>&quot;Timshelvirus droogsarmy&quot;</td><td>0.51</td><td>&quot;-_Duplodnaviria;k_Heunggongvir…</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td></tr><tr><td>&quot;ICTVTaxoChallenge_695160&quot;</td><td>&quot;species&quot;</td><td>5601</td><td>&quot;Badnavirus deltavirgamusae&quot;</td><td>1.0</td><td>&quot;-_Riboviria;k_Pararnavirae;p_A…</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td><td>null</td></tr></tbody></table></div>"
      ],
      "text/plain": [
       "shape: (52_193, 14)\n",
       "┌────────────────────┬─────────┬────────┬───────────────────┬───┬───────┬────────┬───────┬─────────┐\n",
       "│ SequenceID         ┆ rank    ┆ tax_id ┆ taxname           ┆ … ┆ order ┆ family ┆ genus ┆ species │\n",
       "│ ---                ┆ ---     ┆ ---    ┆ ---               ┆   ┆ ---   ┆ ---    ┆ ---   ┆ ---     │\n",
       "│ str                ┆ str     ┆ i64    ┆ str               ┆   ┆ str   ┆ str    ┆ str   ┆ str     │\n",
       "╞════════════════════╪═════════╪════════╪═══════════════════╪═══╪═══════╪════════╪═══════╪═════════╡\n",
       "│ ICTVTaxoChallenge_ ┆ genus   ┆ 7049   ┆ Camvirus          ┆ … ┆ null  ┆ null   ┆ null  ┆ null    │\n",
       "│ 45380              ┆         ┆        ┆                   ┆   ┆       ┆        ┆       ┆         │\n",
       "│ ICTVTaxoChallenge_ ┆ species ┆ 2257   ┆ Emaravirus kiwii  ┆ … ┆ null  ┆ null   ┆ null  ┆ null    │\n",
       "│ 454570             ┆         ┆        ┆                   ┆   ┆       ┆        ┆       ┆         │\n",
       "│ ICTVTaxoChallenge_ ┆ family  ┆ 243    ┆ Steitzviridae     ┆ … ┆ null  ┆ null   ┆ null  ┆ null    │\n",
       "│ 454980             ┆         ┆        ┆                   ┆   ┆       ┆        ┆       ┆         │\n",
       "│ ICTVTaxoChallenge_ ┆ genus   ┆ 1009   ┆ Tetraparvovirus   ┆ … ┆ null  ┆ null   ┆ null  ┆ null    │\n",
       "│ 455533             ┆         ┆        ┆                   ┆   ┆       ┆        ┆       ┆         │\n",
       "│ ICTVTaxoChallenge_ ┆ genus   ┆ 8157   ┆ Phlebovirus       ┆ … ┆ null  ┆ null   ┆ null  ┆ null    │\n",
       "│ 455965             ┆         ┆        ┆                   ┆   ┆       ┆        ┆       ┆         │\n",
       "│ …                  ┆ …       ┆ …      ┆ …                 ┆ … ┆ …     ┆ …      ┆ …     ┆ …       │\n",
       "│ ICTVTaxoChallenge_ ┆ genus   ┆ 11874  ┆ Slopekvirus       ┆ … ┆ null  ┆ null   ┆ null  ┆ null    │\n",
       "│ 693184             ┆         ┆        ┆                   ┆   ┆       ┆        ┆       ┆         │\n",
       "│ ICTVTaxoChallenge_ ┆ species ┆ 3634   ┆ Etapapillomavirus ┆ … ┆ null  ┆ null   ┆ null  ┆ null    │\n",
       "│ 693699             ┆         ┆        ┆ 1                 ┆   ┆       ┆        ┆       ┆         │\n",
       "│ ICTVTaxoChallenge_ ┆ genus   ┆ 9472   ┆ Rotavirus         ┆ … ┆ null  ┆ null   ┆ null  ┆ null    │\n",
       "│ 694175             ┆         ┆        ┆                   ┆   ┆       ┆        ┆       ┆         │\n",
       "│ ICTVTaxoChallenge_ ┆ species ┆ 5351   ┆ Timshelvirus      ┆ … ┆ null  ┆ null   ┆ null  ┆ null    │\n",
       "│ 694718             ┆         ┆        ┆ droogsarmy        ┆   ┆       ┆        ┆       ┆         │\n",
       "│ ICTVTaxoChallenge_ ┆ species ┆ 5601   ┆ Badnavirus        ┆ … ┆ null  ┆ null   ┆ null  ┆ null    │\n",
       "│ 695160             ┆         ┆        ┆ deltavirgamusae   ┆   ┆       ┆        ┆       ┆         │\n",
       "└────────────────────┴─────────┴────────┴───────────────────┴───┴───────┴────────┴───────┴─────────┘"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import re\n",
    "\n",
    "# taxonomyResult_mini = taxonomyResult.rename(\n",
    "#     {'qseqid':\"SequenceID\",'LCA_ncbi_taxid' : 'tax_id',\"taxon_name\" : \"taxname\",\"protein_fragments_in_agreement_with_lca\" : \"score\"}).select(pl.col([\"SequenceID\",\"rank\",\"tax_id\",\"taxname\",\"score\",\"lineage\"]))\n",
    "\n",
    "# # Define the ranks and their prefixes\n",
    "# ranks = {\n",
    "#     'realm': '-_',\n",
    "#     'kingdom': 'k_',\n",
    "#     'phylum': 'p_',\n",
    "#     'class': 'c_',\n",
    "#     'order': 'o_',\n",
    "#     'family': 'f_',\n",
    "#     'genus': 'g_',\n",
    "#     'species': 's_'\n",
    "# }\n",
    "\n",
    "# # Function to extract ranks from lineage\n",
    "# def extract_rank(lineage, rank_prefix):\n",
    "#     pattern = f\"{rank_prefix}([^;]+)\"\n",
    "#     match = re.search(pattern, lineage)\n",
    "#     return match.group(1) if match else \"\"\n",
    "\n",
    "# # Function to set the value of a given rank for each contig as a new column with the value of the taxon_name\n",
    "# def set_rank_value_as_taxon_name(lineage, rank_prefix, taxon_name):\n",
    "#     pattern = f\"{rank_prefix}([^;]+)\"\n",
    "#     match = re.search(pattern, lineage)\n",
    "#     if match:\n",
    "#         return taxon_name\n",
    "#     else:\n",
    "#         return None\n",
    "\n",
    "# # Apply the function to the 'lineage' column for each rank\n",
    "# for rank, prefix in ranks.items():\n",
    "#     taxonomyResult_mini = taxonomyResult_mini.with_columns([\n",
    "#         pl.col('rank').map_elements(lambda x: set_rank_value_as_taxon_name(x, prefix, pl.col('taxname')), return_dtype=pl.Utf8).alias(f'{rank}')\n",
    "#     ])\n",
    "\n",
    "# taxonomyResult_mini"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " Process each unique query"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# # Create a new DataFrame from the results\n",
    "# output_df = pl.DataFrame(results.values())\n",
    "\n",
    "# # Reorder columns to match the desired output format\n",
    "# column_order = ['SequenceID'] + [f'{rank.capitalize()} ({suffix})' for rank, suffix in ranks.items()] + [f'{rank.capitalize()}_score' for rank in ranks.keys()]\n",
    "# output_df = output_df.select(column_order)\n",
    "\n",
    "# # Write the output to a CSV file\n",
    "# output_df.write_csv('taxonomy_output.csv', separator=',')\n",
    "\n",
    "# print(\"Conversion complete. Output saved to 'taxonomy_output.csv'\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
